% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}

\addto\captionsenglish{\renewcommand{\figurename}{Fig. }}
\addto\captionsenglish{\renewcommand{\tablename}{Table }}
\floatname{literal-block}{Listing }



\title{Soprano Documentation}
\date{October 26, 2016}
\release{alpha (Andrews)}
\author{Simone Sturniolo}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@mb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\renewcommand\PYGZsq{\textquotesingle}

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Contents:


\chapter{soprano}
\label{modules:welcome-to-soprano-s-documentation}\label{modules:soprano}\label{modules::doc}

\section{soprano package}
\label{soprano:module-soprano}\label{soprano::doc}\label{soprano:soprano-package}\index{soprano (module)}
Soprano

A Python library to crack crystals
by Simone Sturniolo

Copyright (C) 2016 - Science and Technology Facility Council

Soprano is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Soprano is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}.


\subsection{Subpackages}
\label{soprano:subpackages}

\subsubsection{soprano.analyse package}
\label{soprano.analyse:module-soprano.analyse}\label{soprano.analyse:soprano-analyse-package}\label{soprano.analyse::doc}\index{soprano.analyse (module)}
Contains all modules, classes and functions relevant
to analysing collections after calculations have been performed.


\paragraph{Subpackages}
\label{soprano.analyse:subpackages}

\subparagraph{soprano.analyse.phylogen package}
\label{soprano.analyse.phylogen::doc}\label{soprano.analyse.phylogen:soprano-analyse-phylogen-package}\label{soprano.analyse.phylogen:module-soprano.analyse.phylogen}\index{soprano.analyse.phylogen (module)}
Module containing functions and classes for phylogenetic clustering of
collections.


\subparagraph{Submodules}
\label{soprano.analyse.phylogen:submodules}

\subparagraph{soprano.analyse.phylogen.genes module}
\label{soprano.analyse.phylogen.genes:soprano-analyse-phylogen-genes-module}\label{soprano.analyse.phylogen.genes:module-soprano.analyse.phylogen.genes}\label{soprano.analyse.phylogen.genes::doc}\index{soprano.analyse.phylogen.genes (module)}
Definitions for the various genes used by PhylogenCluster
\index{Gene (class in soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.Gene}\pysiglinewithargsret{\strong{class }\code{soprano.analyse.phylogen.genes.}\bfcode{Gene}}{\emph{name}, \emph{weight=1.0}, \emph{params=\{\}}, \emph{parser=None}, \emph{pair=False}}{}
Bases: \code{object}

A description of a property, a `gene', characterizing a structure, to be
used with a PhylogenCluster. A number of default genes is provided, but
custom ones can be created as well by passing a parser. Only default genes
can be used in a .genefile with the phylogen.py script though.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the gene. Must be one of the existing ones or a
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] custom one (in which case a parser must be provided as
\item[] well). Custom names can't conflict with existing ones
\end{DUlineblock}
\item[] weight (float): weight of the gene to be applied, default is 1.0
\item[] params (dict): additional parameters to be passed to the gene parser
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] function; when not specified, defaults will be used
\end{DUlineblock}
\item[] parser (function\textless{}AtomsCollection, **kwargs\textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} np.array): parser function to be used when defining custom
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genes. Must return a two-dimensional Numpy array
\item[] (axis 0 for the elements of the collection,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] axis 1 for the values of the gene)
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\item[] is\_pair (bool): False if the gene returns a multi dimensional point
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for each structure, True if it only returns pair
\item[] distances. Default is False
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{evaluate() (soprano.analyse.phylogen.genes.Gene method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.Gene.evaluate}\pysiglinewithargsret{\bfcode{evaluate}}{\emph{c}}{}
Evaluate the gene on a given AtomsCollection

\end{fulllineitems}

\index{is\_pair (soprano.analyse.phylogen.genes.Gene attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.Gene.is_pair}\pysigline{\bfcode{is\_pair}}
Whether the gene can only compare a pair of structures or can also
give an absolute value for each structure individually (required for
k-means clustering)

\end{fulllineitems}


\end{fulllineitems}

\index{GeneDictionary (class in soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.GeneDictionary}\pysigline{\strong{class }\code{soprano.analyse.phylogen.genes.}\bfcode{GeneDictionary}}
Bases: \code{object}

Container class holding gene definitions
\index{get\_gene() (soprano.analyse.phylogen.genes.GeneDictionary class method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.GeneDictionary.get_gene}\pysiglinewithargsret{\strong{classmethod }\bfcode{get\_gene}}{\emph{g}}{}
Get the definition for a given gene

\end{fulllineitems}

\index{help() (soprano.analyse.phylogen.genes.GeneDictionary class method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.GeneDictionary.help}\pysiglinewithargsret{\strong{classmethod }\bfcode{help}}{\emph{g}}{}
Get an help string for a given gene

\end{fulllineitems}


\end{fulllineitems}

\index{GeneError}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.GeneError}\pysigline{\strong{exception }\code{soprano.analyse.phylogen.genes.}\bfcode{GeneError}}
Bases: \code{exceptions.Exception}

\end{fulllineitems}

\index{load\_genefile() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.load_genefile}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{load\_genefile}}{\emph{gfile}}{}
Load a gene file and return the (validated) list of genes contained
within.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] gfile (file or str): file to parse
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genelist: a list of genes parsed from the given file, ready to be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] passed to a PhylogenCluster
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{parsegene\_energy() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_energy}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_energy}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_angle() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_angle}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_angle}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_fprint() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_fprint}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_fprint}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_length() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_length}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_length}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_site\_compare() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_site_compare}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_site\_compare}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_site\_reference() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_site_reference}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_site\_reference}}{\emph{c}, \emph{ref=None}}{}
\end{fulllineitems}

\index{parsegene\_hbonds\_totn() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_hbonds_totn}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_hbonds\_totn}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_latt\_abc() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_latt_abc}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_latt\_abc}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_latt\_ang() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_latt_ang}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_latt\_ang}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_latt\_cart() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_latt_cart}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_latt\_cart}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_linkage\_list() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_linkage_list}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_linkage\_list}}{\emph{c}, \emph{size=0}}{}
\end{fulllineitems}

\index{parsegene\_mol\_com() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_mol_com}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_mol\_com}}{\emph{c}, \emph{Z=0}}{}
\end{fulllineitems}

\index{parsegene\_mol\_m() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_mol_m}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_mol\_m}}{\emph{c}, \emph{Z=0}}{}
\end{fulllineitems}

\index{parsegene\_mol\_num() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_mol_num}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_mol\_num}}{\emph{c}}{}
\end{fulllineitems}

\index{parsegene\_mol\_rot() (in module soprano.analyse.phylogen.genes)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.genes:soprano.analyse.phylogen.genes.parsegene_mol_rot}\pysiglinewithargsret{\code{soprano.analyse.phylogen.genes.}\bfcode{parsegene\_mol\_rot}}{\emph{c}, \emph{Z=0}, \emph{twist\_axis=None}, \emph{swing\_plane=None}}{}
\end{fulllineitems}



\subparagraph{soprano.analyse.phylogen.mapping module}
\label{soprano.analyse.phylogen.mapping:soprano-analyse-phylogen-mapping-module}\label{soprano.analyse.phylogen.mapping:module-soprano.analyse.phylogen.mapping}\label{soprano.analyse.phylogen.mapping::doc}\index{soprano.analyse.phylogen.mapping (module)}
2D mapping algorithms
\index{classcond\_principal\_component() (in module soprano.analyse.phylogen.mapping)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.mapping:soprano.analyse.phylogen.mapping.classcond_principal_component}\pysiglinewithargsret{\code{soprano.analyse.phylogen.mapping.}\bfcode{classcond\_principal\_component}}{\emph{p}}{}
\end{fulllineitems}

\index{optimal\_discriminant\_plane() (in module soprano.analyse.phylogen.mapping)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.mapping:soprano.analyse.phylogen.mapping.optimal_discriminant_plane}\pysiglinewithargsret{\code{soprano.analyse.phylogen.mapping.}\bfcode{optimal\_discriminant\_plane}}{\emph{p}}{}
\end{fulllineitems}

\index{standard\_classcond\_component() (in module soprano.analyse.phylogen.mapping)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.mapping:soprano.analyse.phylogen.mapping.standard_classcond_component}\pysiglinewithargsret{\code{soprano.analyse.phylogen.mapping.}\bfcode{standard\_classcond\_component}}{\emph{p}}{}
\end{fulllineitems}

\index{total\_principal\_component() (in module soprano.analyse.phylogen.mapping)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.mapping:soprano.analyse.phylogen.mapping.total_principal_component}\pysiglinewithargsret{\code{soprano.analyse.phylogen.mapping.}\bfcode{total\_principal\_component}}{\emph{p}}{}
\end{fulllineitems}



\subparagraph{soprano.analyse.phylogen.phylogenclust module}
\label{soprano.analyse.phylogen.phylogenclust:module-soprano.analyse.phylogen.phylogenclust}\label{soprano.analyse.phylogen.phylogenclust:soprano-analyse-phylogen-phylogenclust-module}\label{soprano.analyse.phylogen.phylogenclust::doc}\index{soprano.analyse.phylogen.phylogenclust (module)}
Phylogenetic clustering class definitions
\index{PhylogenCluster (class in soprano.analyse.phylogen.phylogenclust)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster}\pysiglinewithargsret{\strong{class }\code{soprano.analyse.phylogen.phylogenclust.}\bfcode{PhylogenCluster}}{\emph{coll}, \emph{genes=None}, \emph{norm\_range=(0.0}, \emph{1.0)}, \emph{norm\_dist=1.0}}{}
Bases: \code{object}

An object that, given an AtomsCollection and a series of ``genes'' and
weights, will build clusters out of the structures in the collection based
on their reciprocal positions as points in a multi-dimensional space
defined by those ``genes''.

Initialize the PhylogenCluster object.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] coll (AtomsCollection): an AtomsCollection containing the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] structures that should be classified.
\item[] This will be copied and frozen for the
\item[] entirety of the life of this instance;
\item[] in order to operate on a modified
\item[] collection, a new PhylogenCluster should
\item[] be created.
\end{DUlineblock}
\item[] genes (list{[}tuple{]}, str, file): list of the genes that should be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] loaded immediately; each gene
\item[] comes in the form of a tuple
\item[] (name (str), weight (float),
\item[] params (dict)). A path or open
\item[] file can also be passed for a
\item[] .gene file, from which the values
\item[] will be loaded.
\end{DUlineblock}
\item[] norm\_range (list{[}float?{]}): ranges to constrain the values of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] single genes in between. Default is
\item[] (0, 1). A value of ``None'' in either
\item[] place can be used to indicate no
\item[] normalization on one or both sides.
\end{DUlineblock}
\item[] norm\_dist (float?): value to normalize distance genes to. These
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] are the genes that only make sense on pairs of
\item[] structures. Their minimum value is always 0.
\item[] This number would become their maximum value,
\item[] or can be set to None to avoid normalization.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{create\_mapping() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping}\pysiglinewithargsret{\bfcode{create\_mapping}}{\emph{method=u'total-principal'}}{}
Return an array of 2-dimensional points representing a reduced
dimensionality mapping of the given genes using the algorithm of
choice. All algorithms are described in {[}W. Siedlecki et al., Patt.
Recog. vol. 21, num. 5, pp. 411 429 (1988){]}.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] method (str): can be one of the following algorithms:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] - total\_principal (default)
\item[] - clafic
\item[] - fukunaga-koontz
\item[] - optimal-discriminant
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_distmat() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat}\pysiglinewithargsret{\bfcode{get\_distmat}}{}{}
Get the distance matrix between structures in the collection,
based on the genes currently in use.

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] distmat (np.ndarray): a (collection.length, collection.length)
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] array, containing the overall distance
\item[] (the norm of all individual gene distances)
\item[] between all pairs of structures.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_genome\_matrices() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices}\pysiglinewithargsret{\bfcode{get\_genome\_matrices}}{}{}
Return the genome matrices in raw form (not normalized).
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genome\_matrix (np.ndarray): a (collection.length,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] collection.length, gene.length)
\item[] array, containing the distances for
\item[] each gene and pair of structures in
\item[] row and column
\end{DUlineblock}
\item[] genome\_legend (list{[}tuple{]}): a list of tuples containing (name,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] length) of the gene fragments in the
\item[] array
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_genome\_matrices\_norm() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm}\pysiglinewithargsret{\bfcode{get\_genome\_matrices\_norm}}{}{}
Return the genome matrices in normalized and weighted form.
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genome\_matrix (np.ndarray): a (collection.length,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] collection.length, gene.length)
\item[] array, containing the distances for
\item[] each gene and pair of structures in
\item[] row and column
\end{DUlineblock}
\item[] genome\_legend (list{[}tuple{]}): a list of tuples containing (name,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] length) of the gene fragments in the
\item[] array
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_genome\_vectors() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors}\pysiglinewithargsret{\bfcode{get\_genome\_vectors}}{}{}
Return the genome vectors in raw form (not normalized).
The vectors refer to genes that allow to define a specific point for
each structure.

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genome\_vectors (np.ndarray): a (collection.length, gene.length)
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] array, containing the whole extent
\item[] of the gene values for each structure
\item[] in the collection on each row
\end{DUlineblock}
\item[] genome\_legend (list{[}tuple{]}): a list of tuples containing (name,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] length) of the gene fragments in the
\item[] array
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_genome\_vectors\_norm() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm}\pysiglinewithargsret{\bfcode{get\_genome\_vectors\_norm}}{}{}
Return the genome vectors in normalized and weighted form.
The vectors refer to genes that allow to define a specific point for
each structure.

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genome\_vectors (np.ndarray): a (collection.length, gene.length)
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] array, containing the whole extent
\item[] of the gene values for each structure
\item[] in the collection on each row
\end{DUlineblock}
\item[] genome\_legend (list{[}tuple{]}): a list of tuples containing (name,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] length) of the gene fragments in the
\item[] array
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_hier\_clusters() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters}\pysiglinewithargsret{\bfcode{get\_hier\_clusters}}{\emph{t}, \emph{method=u'single'}}{}
Get multiple clusters (in the form of a list of collections) based
on the hierarchical clustering methods and the currently set genes.

Calls scipy.cluster.hierarchy.fcluster

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] t (float): minimum distance of separation required to consider
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] two clusters separate. This controls the number of
\item[] clusters: a smaller value will produce more fine
\item[] grained clustering. At the limit, a value smaller than
\item[] the distance between the two closest structures will
\item[] return a cluster for each structure. Remember that the
\item[] `distances' in this case refer to distances between the
\item[] `gene' values attributed to each structure. In other
\item[] words they are a function of the chosen genes,
\item[] normalization conditions and weights employed.
\item[] In addition, the way they are calculated depends on the
\item[] choice of method.
\end{DUlineblock}
\item[] method (str): clustering method to employ. Valid entries are
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] `single', `complete', `weighted' and `average'.
\item[] Refer to Scipy documentation for further details.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] clusters (tuple(list{[}int{]},
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] list{[}slices{]})): list of cluster index for each
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] structure (counting from 1) and
\item[] list of slices defining the
\item[] clusters as formed by hierarchical
\item[] algorithm.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_hier\_tree() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree}\pysiglinewithargsret{\bfcode{get\_hier\_tree}}{\emph{method=u'single'}}{}
Get a tree data structure describing the clustering order of based
on the hierarchical clustering methods and the currently set genes.

Calls scipy.cluster.hierarchy.to\_tree

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] method (str): clustering method to employ. Valid entries are
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] `single', `complete', `weighted' and `average'.
\item[] Refer to Scipy documentation for further details.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] root\_node (ClusterNode): the root node of the tree. Access child
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] members with .left and .right, while .id
\item[] holds the number of the corresponding
\item[] cluster. Refer to Scipy documentation for
\item[] further details.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_kmeans\_clusters() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters}\pysiglinewithargsret{\bfcode{get\_kmeans\_clusters}}{\emph{n}}{}
Get a given number of clusters (in the form of a list of
collections) based on the k-means clustering methods
and the currently set genes.
Warning: this method only works if there are no genes that work only
with pairs of structures - as specific points, and not just distances
between them, are required for this algorithm.

Calls scipy.cluster.vq.kmeans

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] n (int):    the desired number of clusters.
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] clusters (tuple(list{[}int{]},
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] list{[}slices{]})): list of cluster index for each
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] structure (counting from 1) and
\item[] list of slices defining the
\item[] clusters as formed by k-means
\item[] algorithm.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_linkage() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage}\pysiglinewithargsret{\bfcode{get\_linkage}}{\emph{method=u'single'}}{}
Get the linkage matrix between structures in the collection,
based on the genes currently in use. Only used in hierarchical
clustering.

Calls scipy.cluster.hierarchy.linkage.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] method (str): clustering method to employ. Valid entries are
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] `single', `complete', `weighted' and `average'.
\item[] Refer to Scipy documentation for further details.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Z (np.ndarray): linkage matrix for the structures in the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] collection. Refer to Scipy documentation for
\item[] details about the method
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_max\_cluster\_dist() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist}\pysiglinewithargsret{\bfcode{get\_max\_cluster\_dist}}{}{}
Return the maximum possible distance between two clusters

\end{fulllineitems}

\index{load() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load}\pysiglinewithargsret{\strong{static }\bfcode{load}}{\emph{filename}}{}
Load a pickled copy from a given file path

\end{fulllineitems}

\index{save() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save}\pysiglinewithargsret{\bfcode{save}}{\emph{filename}}{}
Simply save a pickled copy to a given file path

\end{fulllineitems}

\index{save\_collection() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection}\pysiglinewithargsret{\bfcode{save\_collection}}{\emph{filename}}{}
Save as pickle the collection bound to this PhylogenCluster.
The calculated genes are also stored in it as arrays for future use.

\end{fulllineitems}

\index{set\_genes() (soprano.analyse.phylogen.phylogenclust.PhylogenCluster method)}

\begin{fulllineitems}
\phantomsection\label{soprano.analyse.phylogen.phylogenclust:soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes}\pysiglinewithargsret{\bfcode{set\_genes}}{\emph{genes}, \emph{load\_arrays=False}}{}
Calculate, store and set a list of genes as used for clustering.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] genes (list{[}soprano.analyse.phylogen.Gene{]},
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] file, str): a list of Genes to calculate and store. A path
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] or open file can also be passed for a .gene
\item[] file, from which the values will be loaded.
\end{DUlineblock}
\end{DUlineblock}
\item[] load\_arrays (bool): try loading the genes as arrays from the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] collection before generating them. Warning:
\item[] if there are arrays named like genes but with
\item[] different contents this can lead to
\item[] unpredictable results.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}



\subsubsection{soprano.calculate package}
\label{soprano.calculate:module-soprano.calculate}\label{soprano.calculate::doc}\label{soprano.calculate:soprano-calculate-package}\index{soprano.calculate (module)}
Contains all modules, classes and functions relevant
to calculating properties of existing structures and collections,
from basic ones to energetic and spectroscopic properties.


\paragraph{Subpackages}
\label{soprano.calculate:subpackages}

\subparagraph{soprano.calculate.gulp package}
\label{soprano.calculate.gulp:module-soprano.calculate.gulp}\label{soprano.calculate.gulp::doc}\label{soprano.calculate.gulp:soprano-calculate-gulp-package}\index{soprano.calculate.gulp (module)}
Classes and functions to carry out calculations using the bindings to GULP
(General Utility Lattice Program), a software providing a lot of useful
calculations with empirical force fields, partial charge calculations, Ewald
summation of Coulombic interactions and more. GULP can be found at:

\href{http://nanochemistry.curtin.edu.au/gulp/}{http://nanochemistry.curtin.edu.au/gulp/}

It needs to be installed on your system to use any of the functionality
provided here.


\subparagraph{Submodules}
\label{soprano.calculate.gulp:submodules}

\subparagraph{soprano.calculate.gulp.charges module}
\label{soprano.calculate.gulp.charges:soprano-calculate-gulp-charges-module}\label{soprano.calculate.gulp.charges:module-soprano.calculate.gulp.charges}\label{soprano.calculate.gulp.charges::doc}\index{soprano.calculate.gulp.charges (module)}
Get charges using GULP
\index{get\_gulp\_charges() (in module soprano.calculate.gulp.charges)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.gulp.charges:soprano.calculate.gulp.charges.get_gulp_charges}\pysiglinewithargsret{\code{soprano.calculate.gulp.charges.}\bfcode{get\_gulp\_charges}}{\emph{s}, \emph{charge\_method=u'eem'}, \emph{save\_charges=True}, \emph{gulp\_command=u'gulp'}, \emph{gulp\_path=None}}{}
Calculate the atomic partial charges using GULP.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] s (ase.Atoms): the structure to calculate the energy of
\item[] charge\_method (Optional{[}str{]}): which method to use for atomic partial
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] charge calculation. Can be any of
\item[] `eem', `qeq' and `pacha'.
\item[] Default is `eem'.
\end{DUlineblock}
\item[] save\_charges (Optional{[}bool{]}): whether to save or not the charges in
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the given ase.Atoms object. Default is
\item[] True.
\end{DUlineblock}
\item[] gulp\_command (Optional{[}str{]}): command required to call the GULP
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] executable.
\end{DUlineblock}
\item[] gulp\_path (Optional{[}str{]}): path where the GULP executable can be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] found. If not present, the GULP command
\item[] will be invoked directly (assuming the
\item[] executable is in the system PATH).
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] charges(np.array(float)): per-atom partial charges
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\subparagraph{soprano.calculate.gulp.w99 module}
\label{soprano.calculate.gulp.w99:soprano-calculate-gulp-w99-module}\label{soprano.calculate.gulp.w99:module-soprano.calculate.gulp.w99}\label{soprano.calculate.gulp.w99::doc}\index{soprano.calculate.gulp.w99 (module)}
Classes and functions for using the W99 force field in GULP.
This force field only applies to organic molecules. More information can be
found in the original paper by Donald E. Williams:

D.E. Williams,
\emph{Improved Intermolecular Force Field for Molecules Containing H, C, N, and O
Atoms, with Application to Nucleoside and Peptide Crystals}  - Journal
of Computational Chemistry, Vol. 22, No. 11, 1154-1166 (2001)
\index{W99Error}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.gulp.w99:soprano.calculate.gulp.w99.W99Error}\pysigline{\strong{exception }\code{soprano.calculate.gulp.w99.}\bfcode{W99Error}}
Bases: \code{exceptions.Exception}

\end{fulllineitems}

\index{find\_w99\_atomtypes() (in module soprano.calculate.gulp.w99)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.gulp.w99:soprano.calculate.gulp.w99.find_w99_atomtypes}\pysiglinewithargsret{\code{soprano.calculate.gulp.w99.}\bfcode{find\_w99\_atomtypes}}{\emph{s}, \emph{force\_recalc=False}}{}
Calculate the W99 force field atom types for a given structure.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] s (ase.Atoms): the structure to calculate the atomtypes on
\item[] force\_recalc (bool): whether to recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present. Default is False.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_w99\_energy() (in module soprano.calculate.gulp.w99)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.gulp.w99:soprano.calculate.gulp.w99.get_w99_energy}\pysiglinewithargsret{\code{soprano.calculate.gulp.w99.}\bfcode{get\_w99\_energy}}{\emph{s}, \emph{charge\_method=u'eem'}, \emph{Etol=1e-06}, \emph{gulp\_command=u'gulp'}, \emph{gulp\_path=None}, \emph{save\_charges=False}}{}
Calculate the W99 force field energy using GULP.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] s (ase.Atoms): the structure to calculate the energy of
\item[] charge\_method (Optional{[}str{]}): which method to use for atomic partial
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] charge calculation. Can be any of
\item[] `eem', `qeq' and `pacha'.
\item[] Default is `eem'.
\end{DUlineblock}
\item[] Etol (Optional{[}float{]}): tolerance on energy for intermolecular
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] potential cutoffs (relative to single
\item[] interaction energy). Default is 1e-6 eV.
\end{DUlineblock}
\item[] gulp\_command (Optional{[}str{]}): command required to call the GULP
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] executable.
\end{DUlineblock}
\item[] gulp\_path (Optional{[}str{]}): path where the GULP executable can be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] found. If not present, the GULP command
\item[] will be invoked directly (assuming the
\item[] executable is in the system PATH).
\end{DUlineblock}
\item[] save\_charges (Optional{[}bool{]}): whether to retrieve also the charges
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] and save them in the Atoms object.
\item[] False by default.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] energy (float): the calculated energy
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\subparagraph{soprano.calculate.xrd package}
\label{soprano.calculate.xrd:soprano-calculate-xrd-package}\label{soprano.calculate.xrd::doc}\label{soprano.calculate.xrd:module-soprano.calculate.xrd}\index{soprano.calculate.xrd (module)}
Classes and functions for simulating X-ray diffraction
spectroscopic results from structures.


\subparagraph{Submodules}
\label{soprano.calculate.xrd:submodules}

\subparagraph{soprano.calculate.xrd.sel\_rules module}
\label{soprano.calculate.xrd.sel_rules::doc}\label{soprano.calculate.xrd.sel_rules:module-soprano.calculate.xrd.sel_rules}\label{soprano.calculate.xrd.sel_rules:soprano-calculate-xrd-sel-rules-module}\index{soprano.calculate.xrd.sel\_rules (module)}
Providing an interface to selection rules for XRD peaks and various
spacegroups.
\index{get\_sel\_rule\_from\_hall() (in module soprano.calculate.xrd.sel\_rules)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.sel_rules:soprano.calculate.xrd.sel_rules.get_sel_rule_from_hall}\pysiglinewithargsret{\code{soprano.calculate.xrd.sel\_rules.}\bfcode{get\_sel\_rule\_from\_hall}}{\emph{h}}{}
Generate a function object that acts as a selection rule for XRD lines
for the given symmetry group expressed in Hall number notation

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] h (int): Hall number of the required spacegroup
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] rule\_func (function\textless{} list\textless{}int\textgreater{} \textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} \textless{}bool\textgreater{}): a function that can be used to test triples of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Miller indices h,k,l to verify whether the
\item[] related plane gives rise or not to a peak
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] RuntimeError: if the database of XRD selection rules or that of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Hall numbers was not properly loaded
\end{DUlineblock}
\item[] ValueError: if the passed argument is invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_sel\_rule\_from\_international() (in module soprano.calculate.xrd.sel\_rules)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.sel_rules:soprano.calculate.xrd.sel_rules.get_sel_rule_from_international}\pysiglinewithargsret{\code{soprano.calculate.xrd.sel\_rules.}\bfcode{get\_sel\_rule\_from\_international}}{\emph{n}, \emph{o=u'all'}}{}
Generate a function object that acts as a selection rule for XRD lines
for the given symmetry group expressed in international number notation

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] n (int): International number of the required spacegroup
\item[] o (Optional{[}int{]}): Sub-option of the required spacegroup
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] rule\_func (function\textless{} list\textless{}int\textgreater{} \textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} \textless{}bool\textgreater{}): a function that can be used to test triples of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Miller indices h,k,l to verify whether the
\item[] related plane gives rise or not to a peak
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] RuntimeError: if the database of XRD selection rules was not properly
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] loaded
\end{DUlineblock}
\item[] ValueError: if some of the passed arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\subparagraph{soprano.calculate.xrd.xrd module}
\label{soprano.calculate.xrd.xrd:module-soprano.calculate.xrd.xrd}\label{soprano.calculate.xrd.xrd::doc}\label{soprano.calculate.xrd.xrd:soprano-calculate-xrd-xrd-module}\index{soprano.calculate.xrd.xrd (module)}
Classes and functions for simulating X-ray diffraction
spectroscopic results from structures.
\index{XRDCalculator (class in soprano.calculate.xrd.xrd)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator}\pysiglinewithargsret{\strong{class }\code{soprano.calculate.xrd.xrd.}\bfcode{XRDCalculator}}{\emph{lambdax=1.54056}, \emph{theta2\_digits=6}, \emph{baseline=0.0}, \emph{peak\_func=None}, \emph{peak\_f\_args=None}}{}
Bases: \code{object}

A class implementing methods for XRD simulations, comparisons and
fittings.

Initialize the XDRCalculator object's main parameters

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] lambdax (Optional{[}float{]}): X-ray wavelength in Angstroms
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] (default is 1.54056 Ang)
\end{DUlineblock}
\item[] theta2\_digits (Optional{[}int{]}): Rounding within which
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] two theta angles (in degrees)
\item[] are considered to be equivalent
\item[] (default is 6 digits) when
\item[] calculating theoretical peaks
\end{DUlineblock}
\end{DUlineblock}
\item[] baseline (Optional{[}float{]}): baseline to use as starting point for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] simulated spectra
\end{DUlineblock}
\item[] peak\_func (Optional{[}function\textless{}float, float, *kargs\textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} \textless{}np.ndarray\textgreater{}{]}): the function used to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] simulate peaks. Should take
\item[] th2 as its first argument,
\item[] peak centre as its second,
\item[] and any number of optional
\item[] arguments. Returns a numpy
\item[] array containing the peak
\item[] shape. Should be able to
\item[] work with numpy arrays as
\item[] input
\end{DUlineblock}
\end{DUlineblock}
\item[] peak\_f\_args (Optional{[}list\textless{}float\textgreater{}{]}): optional arguments for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peak\_func. If no peak\_func
\item[] has been supplied by the
\item[] user, the first value will
\item[] be used as the Gaussian width
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{dataset\_range() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.dataset_range}\pysiglinewithargsret{\bfcode{dataset\_range}}{\emph{xpeaks}, \emph{theta2\_range=(None}, \emph{None)}}{}~\begin{description}
\item[{Restrict the given dataset (XraySpectrum or XraySpectrumData) to}] \leavevmode
only the values that lie within a certain theta2 range.

\end{description}

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks (XraySpectrum or XraySpectrumData): the dataset to modify
\item[] theta2\_range (tuple\textless{}int\textgreater{}): a tuple indicating minimum and maximum
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] of the desired theta2 range (degrees).
\item[] A value of None means no boundary
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks\_restrict (XraySpectrum or XraySpectrumData): the restricted
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] dataset
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the values passed are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{exp\_dataset() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.exp_dataset}\pysiglinewithargsret{\bfcode{exp\_dataset}}{\emph{th2\_axis}, \emph{int\_axis}}{}
Build an experimental dataset as an XraySpectrumData object.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] th2\_axis (np.ndarray): array containing the values for 2*theta
\item[] int\_axis (np.ndarray): array containing the values for intensity
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] exp\_spec (XraySpectrumData): named tuple containing the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] experimental dataset
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the values passed are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{lebail\_fit() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.lebail_fit}\pysiglinewithargsret{\bfcode{lebail\_fit}}{\emph{xpeaks}, \emph{exp\_spec}, \emph{rwp\_tol=0.01}, \emph{max\_iter=100}}{}
Perform a refining on an XraySpectrum object's intensities based on
experimental data with leBail's method.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks (XraySpectrum): object containing the details of the XRD
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peaks
\end{DUlineblock}
\item[] exp\_spec (XraySpectrumData): experimental data, dataset built
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] using xrd\_exp\_dataset
\end{DUlineblock}
\item[] rwp\_tol (Optional{[}float{]}): tolerance on the Rwp error value
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] between two iterations that stops the
\item[] calculation. Default is 1e-2
\end{DUlineblock}
\item[] max\_iter (Optional{[}int{]}): maximum number of iterations to perform
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks\_scaled (XraySpectrum): a new XraySpectrum object, with
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] intensities properly scaled to match
\item[] the experimental data
\end{DUlineblock}
\item[] simul\_spec (np.ndarray): final simulated XRD spectrum
\item[] simul\_peaks (np.ndarray): final simulated spectrum broken by peak
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] contribution along axis 1
\end{DUlineblock}
\item[] rwp (float): the final value of Rwp (fitness of simulated to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] experimental data)
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{peak\_f\_args (soprano.calculate.xrd.xrd.XRDCalculator attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.peak_f_args}\pysigline{\bfcode{peak\_f\_args}}
Additional arguments to be passed to peak\_func

\end{fulllineitems}

\index{peak\_func (soprano.calculate.xrd.xrd.XRDCalculator attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.peak_func}\pysigline{\bfcode{peak\_func}}
The function used to build peaks in simulated spectra

Should be of form peak\_func(theta2, peak\_position, *peak\_f\_args)

\end{fulllineitems}

\index{powder\_peaks() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.powder_peaks}\pysiglinewithargsret{\bfcode{powder\_peaks}}{\emph{atoms=None}, \emph{latt\_abc=None}, \emph{n=1}, \emph{o=u'all'}}{}
Calculate the peaks (without intensities) of a powder
XRD spectrum given either an Atoms object or the lattice in ABC form
and the spacegroup indices to apply the selection rules

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (Optional{[}soprano.Atoms{]}): atoms object to gather lattice
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] and spacegroup information from
\end{DUlineblock}
\item[] latt\_abc (Optional{[}np.ndarray{]}): periodic lattice in ABC form,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Angstroms and radians
\end{DUlineblock}
\end{DUlineblock}
\item[] n (Optional{[}int{]}): International number of the required spacegroup
\item[] o (Optional{[}int{]}): Sub-option of the required spacegroup
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks (XraySpectrum): a named tuple containing the peaks
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] with theta2, corresponding hkl indices,
\item[] a unique hkl tuple for each peak,
\item[] inverse reciprocal lattice distances,
\item[] intensities and wavelength
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{set\_peak\_func() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.set_peak_func}\pysiglinewithargsret{\bfcode{set\_peak\_func}}{\emph{peak\_func=None}, \emph{peak\_f\_args=None}}{}
Set a new peak\_func for this XDRCalculator. If no new function is
passed, reset the default Gaussian function.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peak\_func (Optional{[}function\textless{}float, float, *kargs\textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} \textless{}np.ndarray\textgreater{}{]}): the function used to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] simulate peaks. Should
\item[] take th2 as its first
\item[] argument, peak centre as
\item[] its second, and any
\item[] number of optional
\item[] arguments. Returns a
\item[] numpy array containing
\item[] the peak shape. Should
\item[] be able to work with
\item[] numpy arrays as input
\end{DUlineblock}
\end{DUlineblock}
\item[] peak\_f\_args (Optional{[}list\textless{}float\textgreater{}{]}): optional arguments for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peak\_func. If no peak\_func
\item[] has been supplied by the
\item[] user, the first value will
\item[] be used as the Gaussian
\item[] width
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{spec\_simul() (soprano.calculate.xrd.xrd.XRDCalculator method)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XRDCalculator.spec_simul}\pysiglinewithargsret{\bfcode{spec\_simul}}{\emph{xpeaks}, \emph{th2\_axis}}{}
Simulate an XRD spectrum given positions of peaks, intensities,
baseline, and a peak function (a Gaussian by default).

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] xpeaks (XraySpectrum): object containing the details of the XRD
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peaks
\end{DUlineblock}
\item[] th2\_axis (np.ndarray): theta2 axis points on which the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] spectrum should be simulated
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] simul\_spec (XraySpectrumData): simulated XRD spectrum
\item[] simul\_peaks (np.ndarray): simulated spectrum intensities broken by
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] peak contribution along axis 1
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}

\index{XraySpectrum (class in soprano.calculate.xrd.xrd)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum}\pysiglinewithargsret{\strong{class }\code{soprano.calculate.xrd.xrd.}\bfcode{XraySpectrum}}{\emph{theta2}, \emph{hkl}, \emph{hkl\_unique}, \emph{invd}, \emph{intensity}, \emph{lambdax}}{}
Bases: \code{tuple}
\index{hkl (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.hkl}\pysigline{\bfcode{hkl}}
Alias for field number 1

\end{fulllineitems}

\index{hkl\_unique (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.hkl_unique}\pysigline{\bfcode{hkl\_unique}}
Alias for field number 2

\end{fulllineitems}

\index{intensity (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.intensity}\pysigline{\bfcode{intensity}}
Alias for field number 4

\end{fulllineitems}

\index{invd (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.invd}\pysigline{\bfcode{invd}}
Alias for field number 3

\end{fulllineitems}

\index{lambdax (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.lambdax}\pysigline{\bfcode{lambdax}}
Alias for field number 5

\end{fulllineitems}

\index{theta2 (soprano.calculate.xrd.xrd.XraySpectrum attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrum.theta2}\pysigline{\bfcode{theta2}}
Alias for field number 0

\end{fulllineitems}


\end{fulllineitems}

\index{XraySpectrumData (class in soprano.calculate.xrd.xrd)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrumData}\pysiglinewithargsret{\strong{class }\code{soprano.calculate.xrd.xrd.}\bfcode{XraySpectrumData}}{\emph{theta2}, \emph{intensity}}{}
Bases: \code{tuple}
\index{intensity (soprano.calculate.xrd.xrd.XraySpectrumData attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrumData.intensity}\pysigline{\bfcode{intensity}}
Alias for field number 1

\end{fulllineitems}

\index{theta2 (soprano.calculate.xrd.xrd.XraySpectrumData attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.calculate.xrd.xrd:soprano.calculate.xrd.xrd.XraySpectrumData.theta2}\pysigline{\bfcode{theta2}}
Alias for field number 0

\end{fulllineitems}


\end{fulllineitems}



\subsubsection{soprano.collection package}
\label{soprano.collection:soprano-collection-package}\label{soprano.collection::doc}\label{soprano.collection:module-soprano.collection}\index{soprano.collection (module)}
Contains all modules, classes and functions relevant
to handling, loading or randomly generating
collections of structures.


\paragraph{Subpackages}
\label{soprano.collection:subpackages}

\subparagraph{soprano.collection.generate package}
\label{soprano.collection.generate:module-soprano.collection.generate}\label{soprano.collection.generate:soprano-collection-generate-package}\label{soprano.collection.generate::doc}\index{soprano.collection.generate (module)}
This module contains generators meant to produce AtomsCollections based
on different criteria.


\subparagraph{Submodules}
\label{soprano.collection.generate:submodules}

\subparagraph{soprano.collection.generate.airss module}
\label{soprano.collection.generate.airss::doc}\label{soprano.collection.generate.airss:module-soprano.collection.generate.airss}\label{soprano.collection.generate.airss:soprano-collection-generate-airss-module}\index{soprano.collection.generate.airss (module)}
Bindings for AIRSS Buildcell program for random structure generation
\index{airssGen() (in module soprano.collection.generate.airss)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.generate.airss:soprano.collection.generate.airss.airssGen}\pysiglinewithargsret{\code{soprano.collection.generate.airss.}\bfcode{airssGen}}{\emph{input\_file}, \emph{n=100}, \emph{buildcell\_command=u'buildcell'}, \emph{buildcell\_path=None}}{}
Generator function binding to AIRSS' Buildcell.

This function searches for a buildcell executable and uses it to
generate multiple new Atoms structures for a collection.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] input\_file (str or file): the .cell file with appropriate comments
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] specifying the details of buildcell's
\item[] construction work.
\end{DUlineblock}
\item[] n (int): number of structures to generate. If set to None the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] generator goes on indefinitely.
\end{DUlineblock}
\item[] buildcell\_command (str): command required to call the buildcell
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] executable.
\end{DUlineblock}
\item[] buildcell\_path (str): path where the buildcell executable can be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] found. If not present, the buildcell command
\item[] will be invoked directly (assuming the
\item[] executable is in the system PATH).
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] airssGenerator (generator): an iterable object that yields structures
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] created by buildcell.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\subparagraph{soprano.collection.generate.linspace module}
\label{soprano.collection.generate.linspace::doc}\label{soprano.collection.generate.linspace:soprano-collection-generate-linspace-module}\label{soprano.collection.generate.linspace:module-soprano.collection.generate.linspace}\index{soprano.collection.generate.linspace (module)}
Generator producing structures interpolated between two extremes
\index{linspaceGen() (in module soprano.collection.generate.linspace)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.generate.linspace:soprano.collection.generate.linspace.linspaceGen}\pysiglinewithargsret{\code{soprano.collection.generate.linspace.}\bfcode{linspaceGen}}{\emph{struct\_0}, \emph{struct\_1}, \emph{steps=10}, \emph{periodic=False}}{}
Generator function to create multiple structures with positions
interpolated linearly between two extremes.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] struct\_0 (ase.Atoms): the starting structure
\item[] struct\_1 (ase.Atoms): the final structure. The atoms should be in the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] same order as the ones in struct\_0
\end{DUlineblock}
\item[] steps (Optional{[}int{]}): number of interpolated steps to produce
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] (extremes included). Default is 10
\end{DUlineblock}
\item[] periodic (Optional{[}bool{]}): if True the interpolation will take into
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] account periodic boundaries and interpolate
\item[] between positions in struct\_0 and the
\item[] closest periodic copy of positions in
\item[] struct\_1. By default set to False
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] linspaceGenerator (generator): an iterator object that yields
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] structures created by linear
\item[] interpolation.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\subparagraph{soprano.collection.generate.rattle module}
\label{soprano.collection.generate.rattle::doc}\label{soprano.collection.generate.rattle:soprano-collection-generate-rattle-module}\label{soprano.collection.generate.rattle:module-soprano.collection.generate.rattle}\index{soprano.collection.generate.rattle (module)}
Generator producing structures rattled of a given amount
\index{rattleGen() (in module soprano.collection.generate.rattle)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.generate.rattle:soprano.collection.generate.rattle.rattleGen}\pysiglinewithargsret{\code{soprano.collection.generate.rattle.}\bfcode{rattleGen}}{\emph{struct}, \emph{amplitude=0.01}, \emph{n=100}, \emph{method=u'uniform'}}{}
Generator function to create multiple structures by randomly displacing
atoms of a given amount.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] struct (ase.Atoms): the starting structure to randomize
\item[] amplitude (float or np.ndarray): the amplitude of the random
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] displacement. Can be a single float
\item[] for all atoms, a 1D numpy array of
\item[] length N (N being the number of
\item[] atoms, one value each) or a 2D numpy
\item[] array of shape (N,3) (one value for
\item[] each dimension).
\item[] These values are used as interval for
\item[] uniform random numbers and as stdev
\item[] for normal random numbers
\end{DUlineblock}
\item[] n (int): maximum number of structures to generate. If set to None will
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] generate infinite structures
\end{DUlineblock}
\item[] method (str): must be either `uniform' or `normal'. In the first case
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the rattling will be a uniform random number between
\item[] +amplitude and -amplitude. In the second case it will be
\item[] a gaussian random number with +amplitude standard
\item[] deviation.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] rattleGenerator (generator): an iterator that yields copies of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] base structure with randomly displaced
\item[] atoms.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}



\paragraph{Submodules}
\label{soprano.collection:submodules}

\subparagraph{soprano.collection.collection module}
\label{soprano.collection.collection:module-soprano.collection.collection}\label{soprano.collection.collection::doc}\label{soprano.collection.collection:soprano-collection-collection-module}\index{soprano.collection.collection (module)}
Definition of the Collection class.

It handles multiple Atoms ASE objects and mirrors in this sense the structure
of the Atoms object itself.
\index{AtomsCollection (class in soprano.collection.collection)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection}\pysiglinewithargsret{\strong{class }\code{soprano.collection.collection.}\bfcode{AtomsCollection}}{\emph{structures={[}{]}}, \emph{info=\{\}}, \emph{cell\_reduce=False}, \emph{progress=False}}{}
Bases: \code{object}

AtomsCollection object.

An AtomsCollection represents a group of ASE Atoms objects.
It handles them together, can perform mass operations on them, and stores
arrays of informations related to them.

Initialize the AtomsCollection

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] structures (list{[}str{]} or list{[}ase.Atoms{]}): list of file names or
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Atoms that will form
\item[] the collection
\end{DUlineblock}
\item[] info (dict): dictionary of general information to attach
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] to this collection
\end{DUlineblock}
\item[] cell\_reduce (bool): if True, perform a Niggli cell reduction on
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] all loaded structures
\end{DUlineblock}
\item[] progress (bool): visualize a progress bar for the loading process
\end{DUlineblock}
\end{DUlineblock}
\index{all (soprano.collection.collection.AtomsCollection attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.all}\pysigline{\bfcode{all}}
\end{fulllineitems}

\index{chunkify() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.chunkify}\pysiglinewithargsret{\bfcode{chunkify}}{\emph{chunk\_size=None}, \emph{chunk\_n=None}}{}
Split this collection into multiple collections based on either
size or number of chunks.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] chunk\_size (Optional{[}int{]}): maximum size of a generated chunk
\item[] chunk\_n (Optional{[}int{]}): number of chunks to generate
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] chunks (list{[}AtomsCollection{]}): a list of the generated chunks
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{filter() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.filter}\pysiglinewithargsret{\bfcode{filter}}{\emph{filter\_func}}{}
Return a collection composed only of the elements for which a given
filter function returns True.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] filter\_func (function\textless{}Atoms\textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} bool): filter function. Should take an
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Atoms object and return a boolean
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] filtered (AtomsCollection): the filtered version of the collection
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_array() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.get_array}\pysiglinewithargsret{\bfcode{get\_array}}{\emph{name}, \emph{copy=True}}{}
Get a copy of an array of given name (or a reference if copy=False)

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the array to retrieve.
\item[] copy (bool): if the array should be copied or a reference should
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] be returned instead.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] array (np.ndarray): the requested array
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{has() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.has}\pysiglinewithargsret{\bfcode{has}}{\emph{name}}{}
Check if array of given name exists

\end{fulllineitems}

\index{length (soprano.collection.collection.AtomsCollection attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.length}\pysigline{\bfcode{length}}
\end{fulllineitems}

\index{load() (soprano.collection.collection.AtomsCollection static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.load}\pysiglinewithargsret{\strong{static }\bfcode{load}}{\emph{filename}}{}
Load a pickled copy from a given file path

\end{fulllineitems}

\index{run\_calculators() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.run_calculators}\pysiglinewithargsret{\bfcode{run\_calculators}}{\emph{properties=None}, \emph{system\_changes=None}}{}
Run all previously set ASE calculators.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] properties (list{[}str{]}): list of properties to calculate (depends
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] on type of Calculator used)
\end{DUlineblock}
\item[] system\_changes (list{[}str{]}): list of changes to the structure
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] since the last calculation. Can be
\item[] any combination of these five:
\item[] `positions', `numbers', `cell',
\item[] `pbc', `initial\_charges' and
\item[] `initial\_magmoms'.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{save() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.save}\pysiglinewithargsret{\bfcode{save}}{\emph{filename}}{}
Simply save a pickled copy to a given file path

\end{fulllineitems}

\index{set\_array() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.set_array}\pysiglinewithargsret{\bfcode{set\_array}}{\emph{name}, \emph{a}, \emph{dtype=None}, \emph{shape=None}, \emph{args=\{\}}}{}
Add or modify an array of data related to the Atoms objects
in this collection.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the array to operate on.
\item[] a (np.ndarray or function\textless{}Atoms, **kwargs\textgreater{}
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] =\textgreater{} Any): the data to assign to the array (must
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] be same length as the collection) or
\item[] a function that takes an Atoms object
\item[] as the first argument and returns a
\item[] value. This will be mapped over the
\item[] structures to create the array.
\end{DUlineblock}
\end{DUlineblock}
\item[] dtype (type): type to cast the values of the array to.
\item[] shape (tuple {[}int{]}): shape of each entry of the array. Will be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] checked if provided.
\end{DUlineblock}
\item[] args (dict): named arguments to pass to the function provided
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] as a. Will be ignored if an array is passed instead.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{set\_calculators() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.set_calculators}\pysiglinewithargsret{\bfcode{set\_calculators}}{\emph{calctype}, \emph{labels=None}, \emph{params=\{\}}}{}
Set an ASE calculator on each structure in the collection,
and set said calculator's parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] calctype (ASE Calculator type): the type of calculator
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] to instantiate.
\end{DUlineblock}
\item[] labels (Optional{[}list{[}str{]}{]}): names to use for the calculators'
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] files. If not present, random
\item[] generated names are used.
\end{DUlineblock}
\item[] params (Optional{[}dict{]}): parameters of the calculator to set.
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{sorted\_byarray() (soprano.collection.collection.AtomsCollection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.collection.collection:soprano.collection.collection.AtomsCollection.sorted_byarray}\pysiglinewithargsret{\bfcode{sorted\_byarray}}{\emph{name}, \emph{reverse=False}}{}
Return a copy of this collection sorted by a given array.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the array to use for the sorting
\item[] reverse (Optional{[}bool{]}): reverse order of sorting (max to min)
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] sorted (AtomsCollection): a sorted copy of the collection
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}



\subsubsection{soprano.hpc package}
\label{soprano.hpc:module-soprano.hpc}\label{soprano.hpc:soprano-hpc-package}\label{soprano.hpc::doc}\index{soprano.hpc (module)}
Classes and functions useful to run calculations on huge cluster systems
(High Performance Computation). To be used with care - these are liable to
fail if some specific architecture has quirks that are not accounted for!


\paragraph{Subpackages}
\label{soprano.hpc:subpackages}

\subparagraph{soprano.hpc.submitter package}
\label{soprano.hpc.submitter:soprano-hpc-submitter-package}\label{soprano.hpc.submitter::doc}\label{soprano.hpc.submitter:module-soprano.hpc.submitter}\index{soprano.hpc.submitter (module)}
Classes and functions required for processes that automatically submit jobs
to a queueing system working in the background.

These can be launched interactively from the command line. In order to do
that:
\begin{enumerate}
\item {} 
write your own implementation of a submitter class by inheriting from
soprano.hpc.submitter.Submitter or use one of the provided ones;

\item {} 
write an input file in which you simply create an instance of said class
and set up its parameters (ideally by calling the set\_parameters method);

\item {} 
launch that submitter from the command line with the following command:
\begin{quote}

\code{python -m soprano.hpc.submitter start \textless{}filename\textgreater{}}
\end{quote}

\end{enumerate}

You can have multiple submitter instances, even of different types, defined
in the same file: in that case you will need to use the -n option to specify
which one you want to launch (the name you need to use is the name of the
\emph{variable} you stored the instance in). If you are working on remote login and
you want to prevent the submitter from being terminated upon exiting your
session use the -nohup option.
To list which submitters from a given file are running, and how long have they
been running for, just use:
\begin{quote}

\code{python -m soprano.hpc.submitter list \textless{}filename\textgreater{}}
\end{quote}

Similarly, you can stop a running submitter with:
\begin{quote}

\code{python -m soprano.hpc.submitter stop \textless{}filename\textgreater{}}
\end{quote}

Submitters have a `name' property and will save a \textless{}name\textgreater{}.log file in which any
output from their run can be stored.


\subparagraph{Subpackages}
\label{soprano.hpc.submitter:subpackages}

\subparagraph{soprano.hpc.submitter.debug package}
\label{soprano.hpc.submitter.debug:module-soprano.hpc.submitter.debug}\label{soprano.hpc.submitter.debug:soprano-hpc-submitter-debug-package}\label{soprano.hpc.submitter.debug::doc}\index{soprano.hpc.submitter.debug (module)}
Functions useful for debugging QueueInterface and Submitters. These
provide a `fake' queue that executes basic jobs with artificial delays
in order to simulate an environment similar to what can be found on an HPC
machine.


\subparagraph{Submodules}
\label{soprano.hpc.submitter.debug:submodules}

\subparagraph{soprano.hpc.submitter.debug.debugqueue module}
\label{soprano.hpc.submitter.debug.debugqueue:soprano-hpc-submitter-debug-debugqueue-module}\label{soprano.hpc.submitter.debug.debugqueue::doc}\label{soprano.hpc.submitter.debug.debugqueue:module-soprano.hpc.submitter.debug.debugqueue}\index{soprano.hpc.submitter.debug.debugqueue (module)}
Definition of a fake QueueInterface class, useful for debugging Submitters.
\index{DebugQueueInterface (class in soprano.hpc.submitter.debug.debugqueue)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.debug.debugqueue:soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface}\pysiglinewithargsret{\strong{class }\code{soprano.hpc.submitter.debug.debugqueue.}\bfcode{DebugQueueInterface}}{\emph{dt=0.1}}{}
Bases: {\hyperref[soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface]{\emph{\code{soprano.hpc.submitter.queues.QueueInterface}}}}

DebugQueueInterface object

A class meant to emulate a QueueInterface while doing absolutely nothing
of what it does. Jobs are simply stored locally, there's a fixed waiting
time, and are then executed. Ideally they should be simple, quick stuff
(like an echo command). No guarantees for actually long jobs.

In the submitted script a syntax for additional variables is allowed,
similar to real queue systems. These follow the convention of many engines
of having to start with \#\$. For example

\#\$ WAIT 10

means the job will be put in a ``wait'' state for 10 seconds. The currently
available variables are:

WAIT - specify how long the job has to stay in a ``wait'' state. If two
values are provided, these are considered bounds for a random number
RUN - same as above, but for the running state. This has no bearing on the
\emph{actual} running time (it's suggested that it's something very quick)

Initialize the DebugQueueInterface.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] dt (float): frequency with which the queue status is updated
\end{DUlineblock}
\end{DUlineblock}
\index{kill() (soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.debug.debugqueue:soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface.kill}\pysiglinewithargsret{\bfcode{kill}}{\emph{job\_id}}{}
Kill the job with the given ID

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job\_id (str): ID of the job to kill
\item[] 
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{list() (soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.debug.debugqueue:soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface.list}\pysiglinewithargsret{\bfcode{list}}{}{}
List all jobs found in the queue

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] jobs (dict): a dict of jobs classified by ID containing all info
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] that can be matched through list\_outre
\item[] 
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{submit() (soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.debug.debugqueue:soprano.hpc.submitter.debug.debugqueue.DebugQueueInterface.submit}\pysiglinewithargsret{\bfcode{submit}}{\emph{script}, \emph{cwd=None}}{}
Submit a job to the queue.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] script (str): content of the submission script
\item[] cwd (Optional{[}str{]}): path to the desired working directory
\item[] 
\end{DUlineblock}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job\_id (str): the job ID assigned by the queue system and parsed
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] with sub\_outre
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{Submodules}
\label{soprano.hpc.submitter:submodules}

\subparagraph{soprano.hpc.submitter.castep module}
\label{soprano.hpc.submitter.castep:module-soprano.hpc.submitter.castep}\label{soprano.hpc.submitter.castep::doc}\label{soprano.hpc.submitter.castep:soprano-hpc-submitter-castep-module}\index{soprano.hpc.submitter.castep (module)}
Definition of CastepSubmitter class.

A basic ``rolling'' submitter for Castep calculations, grabbing from one folder
and depositing results in another.
\index{CastepSubmitter (class in soprano.hpc.submitter.castep)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter}\pysiglinewithargsret{\strong{class }\code{soprano.hpc.submitter.castep.}\bfcode{CastepSubmitter}}{\emph{name}, \emph{queue}, \emph{submit\_script}, \emph{max\_jobs=4}, \emph{check\_time=10}, \emph{max\_time=3600}, \emph{temp\_folder=None}}{}
Bases: {\hyperref[soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter]{\emph{\code{soprano.hpc.submitter.submit.Submitter}}}}

Initialize the Submitter object

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name to be used for this Submitter (two Submitters
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] with the same name can't be launched in the same
\item[] working directory)
\end{DUlineblock}
\item[] queue (QueueInterface): object describing the properties of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] interface to the queue system in use
\end{DUlineblock}
\item[] submit\_script (str): text of the script to use when submitting a
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job to the queue. All tags of the form \textless{}name\textgreater{}
\item[] will be replaced with the job's name, and all
\item[] similar tags of the form \textless{}{[}arg{]}\textgreater{} will be
\item[] replaced if the argument name is present in
\item[] the job's args dictionary
\end{DUlineblock}
\item[] max\_jobs (Optional{[}int{]}): maximum number of jobs to submit at a
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] given time. Default is 4
\end{DUlineblock}
\item[] check\_time (Optional{[}float{]}): time in seconds between consecutive
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] checks for the queue status and
\item[] attempts to submit new jobs. Default
\item[] is 10
\end{DUlineblock}
\item[] max\_time (Optional{[}float{]}): time in seconds the Submitter will run
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for before shutting down. If set to
\item[] zero the thread won't stop until
\item[] killed with Submitter.stop.
\end{DUlineblock}
\item[] temp\_folder (Optional{[}str{]}): where to store the temporary folders
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for the calculations. By default it's
\item[] the system's tmp/ folder, but might
\item[] be changed if there's a need because
\item[] of writing permissions.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{finish\_job() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.finish_job}\pysiglinewithargsret{\bfcode{finish\_job}}{\emph{name}, \emph{args}, \emph{folder}}{}
Save required output files to the output folder

\end{fulllineitems}

\index{finish\_run() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.finish_run}\pysiglinewithargsret{\bfcode{finish\_run}}{}{}
Try removing the temporary keywords directory

\end{fulllineitems}

\index{next\_job() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.next_job}\pysiglinewithargsret{\bfcode{next\_job}}{}{}
Grab the next job from folder\_in

\end{fulllineitems}

\index{set\_parameters() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.set_parameters}\pysiglinewithargsret{\bfcode{set\_parameters}}{\emph{folder\_in, folder\_out, castep\_command, castep\_path=None, copy\_extensions={[}u'.castep'{]}, pspot\_files={[}{]}, dryrun\_test=False}}{}
Set the parameters of the CASTEP Submitter

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] folder\_in (str): path of the folder to extract cell files from
\item[] folder\_out (str): path of the folder where the results will be
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] saved
\end{DUlineblock}
\item[] castep\_command (str): command used to call the CASTEP executable
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] on this system
\end{DUlineblock}
\item[] castep\_path (Optional{[}str{]}): folder where the CASTEP executable is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] located (if not part of the system
\item[] PATH)
\end{DUlineblock}
\item[] pspot\_files (Optional{[}list{[}str{]}{]}): additional pseudopotential
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] files to be copied in the input
\item[] temporary folders
\end{DUlineblock}
\item[] copy\_extensions (Optional{[}list{[}str{]}{]}): extensions of output files
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] to copy to the output
\item[] folder (by default only
\item[] .castep file)
\end{DUlineblock}
\item[] dryrun\_test (Optional{[}bool{]}): run a dryrun test on files before
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] actually running the calculation.
\item[] Off by default.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{setup\_job() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.setup_job}\pysiglinewithargsret{\bfcode{setup\_job}}{\emph{name}, \emph{args}, \emph{folder}}{}
Copy files to temporary folder to prepare for execution

\end{fulllineitems}

\index{start\_run() (soprano.hpc.submitter.castep.CastepSubmitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.castep:soprano.hpc.submitter.castep.CastepSubmitter.start_run}\pysiglinewithargsret{\bfcode{start\_run}}{}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.hpc.submitter.queues module}
\label{soprano.hpc.submitter.queues::doc}\label{soprano.hpc.submitter.queues:module-soprano.hpc.submitter.queues}\label{soprano.hpc.submitter.queues:soprano-hpc-submitter-queues-module}\index{soprano.hpc.submitter.queues (module)}
Definition of QueueInterface class.
\index{QueueInterface (class in soprano.hpc.submitter.queues)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface}\pysiglinewithargsret{\strong{class }\code{soprano.hpc.submitter.queues.}\bfcode{QueueInterface}}{\emph{sub\_cmd}, \emph{list\_cmd}, \emph{kill\_cmd}, \emph{sub\_outre}, \emph{list\_outre}}{}
Bases: \code{object}

QueueInterface object

A class meant to simplify interfacing in a basic way
with a Queue system. Contains commands to submit to the queue, list the
job IDs, and kill them if necessary. Will contain Regexps to parse for IDs
and additional information as returned upon submission and listing.
It is important that the regular expressions used employ NAMED GROUPS to
parse the various fields. In particular, a job\_id group must ALWAYS be
present.
The class also provides some static variables implementing standard
interfaces for common queueing systems. These can be retrieved by using
QueueInterface.\textless{}NAME\textgreater{}. The currently implemented names are the following:
\begin{itemize}
\item {} 
LSF (IBM's managing system, using the command bsub)

\item {} \begin{description}
\item[{GridEngine (Sun's managing system, also available in an open version,}] \leavevmode
using the command qsub)

\end{description}

\end{itemize}

Initialize the QueueInterface.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] sub\_cmd (str): command used to submit a script to the queue
\item[] list\_cmd (str): command used to list all queued jobs for the user
\item[] kill\_cmd (str): command used to kill a job given its id
\item[] sub\_outre (str): regular expression used to parse the output of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] sub\_cmd. Must contain at least a job\_id named
\item[] group
\end{DUlineblock}
\item[] list\_outre (str): regular expression used to parse the output of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] list\_cmd. Must contain at least a job\_id named
\item[] group
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{GridEngine() (soprano.hpc.submitter.queues.QueueInterface class method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface.GridEngine}\pysiglinewithargsret{\strong{classmethod }\bfcode{GridEngine}}{}{}
\end{fulllineitems}

\index{LSF() (soprano.hpc.submitter.queues.QueueInterface class method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface.LSF}\pysiglinewithargsret{\strong{classmethod }\bfcode{LSF}}{}{}
\end{fulllineitems}

\index{kill() (soprano.hpc.submitter.queues.QueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface.kill}\pysiglinewithargsret{\bfcode{kill}}{\emph{job\_id}}{}
Kill the job with the given ID

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job\_id (str): ID of the job to kill
\item[] 
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{list() (soprano.hpc.submitter.queues.QueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface.list}\pysiglinewithargsret{\bfcode{list}}{}{}
List all jobs found in the queue

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] jobs (dict): a dict of jobs classified by ID containing all info
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] that can be matched through list\_outre
\item[] 
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{submit() (soprano.hpc.submitter.queues.QueueInterface method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.queues:soprano.hpc.submitter.queues.QueueInterface.submit}\pysiglinewithargsret{\bfcode{submit}}{\emph{script}, \emph{cwd=None}}{}
Submit a job to the queue.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] script (str): content of the submission script
\item[] cwd (Optional{[}str{]}): path to the desired working directory
\item[] 
\end{DUlineblock}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job\_id (str): the job ID assigned by the queue system and parsed
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] with sub\_outre
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.hpc.submitter.submit module}
\label{soprano.hpc.submitter.submit:soprano-hpc-submitter-submit-module}\label{soprano.hpc.submitter.submit::doc}\label{soprano.hpc.submitter.submit:module-soprano.hpc.submitter.submit}\index{soprano.hpc.submitter.submit (module)}
Definition of Submitter class

Base class for all Submitters to inherit from.
\index{Submitter (class in soprano.hpc.submitter.submit)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter}\pysiglinewithargsret{\strong{class }\code{soprano.hpc.submitter.submit.}\bfcode{Submitter}}{\emph{name}, \emph{queue}, \emph{submit\_script}, \emph{max\_jobs=4}, \emph{check\_time=10}, \emph{max\_time=3600}, \emph{temp\_folder=None}}{}
Bases: \code{object}

Submitter object

Template to derive all specialised Submitters. These are meant to generate,
submit and post-process any number of jobs on a queueing system in the form
of a background process running on a head node. It implements
methods that should be mostly overridden by the child classes.
Six methods define its core behaviour:
\begin{enumerate}
\item {} 
next\_job is the function that outputs the specification for each new job
to submit. The specification should be a dict with two members, `name'
(a string) and `args' (ideally a dict). If no more jobs are available
it should return None;

\item {} 
setup\_job takes as arguments name, args and folder (a temporary one
created independently) and is supposed to generate the input files
for the job before submission. It returns a boolean, confirming that
the setup went well; if False, the job will be skipped;

\item {} 
check\_job takes as arguments job ID, name, args and folder and should
return a bool confirmation of whether the job has finished or not. By
default it simply checks whether the job is still listed in the queue,
however other checks can be implemented in its place;

\item {} 
finish\_job takes as arguments name, args and folder and takes care of
the post processing once a job is complete. Here meaningful data should
be extracted and useful files copied to permament locations, as the
temporary folder will be deleted immediately afterwards. It returns
nothing;

\item {} 
start\_run takes no arguments, executes at the beginning of a run;

\item {} 
finish\_run takes no arguments, executes at the end of a run.

\end{enumerate}

In addition, the Submitter takes a template launching script which can
be tagged with keywords, mainly \textless{}name\textgreater{} for the job name or any other
arguments present in args. These will be replaced with the appropriate
values when the script is submitted.

Initialize the Submitter object

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name to be used for this Submitter (two Submitters
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] with the same name can't be launched in the same
\item[] working directory)
\end{DUlineblock}
\item[] queue (QueueInterface): object describing the properties of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] interface to the queue system in use
\end{DUlineblock}
\item[] submit\_script (str): text of the script to use when submitting a
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] job to the queue. All tags of the form \textless{}name\textgreater{}
\item[] will be replaced with the job's name, and all
\item[] similar tags of the form \textless{}{[}arg{]}\textgreater{} will be
\item[] replaced if the argument name is present in
\item[] the job's args dictionary
\end{DUlineblock}
\item[] max\_jobs (Optional{[}int{]}): maximum number of jobs to submit at a
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] given time. Default is 4
\end{DUlineblock}
\item[] check\_time (Optional{[}float{]}): time in seconds between consecutive
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] checks for the queue status and
\item[] attempts to submit new jobs. Default
\item[] is 10
\end{DUlineblock}
\item[] max\_time (Optional{[}float{]}): time in seconds the Submitter will run
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for before shutting down. If set to
\item[] zero the thread won't stop until
\item[] killed with Submitter.stop.
\end{DUlineblock}
\item[] temp\_folder (Optional{[}str{]}): where to store the temporary folders
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for the calculations. By default it's
\item[] the system's tmp/ folder, but might
\item[] be changed if there's a need because
\item[] of writing permissions.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{check\_job() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.check_job}\pysiglinewithargsret{\bfcode{check\_job}}{\emph{job\_id}, \emph{name}, \emph{args}, \emph{folder}}{}
Checks if given job is complete or not

\end{fulllineitems}

\index{finish\_job() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.finish_job}\pysiglinewithargsret{\bfcode{finish\_job}}{\emph{name}, \emph{args}, \emph{folder}}{}
Performs completiion operations on the job. At this point any
relevant output files should be copied from `folder' to their final
destination as the temporary folder itself will be deleted immediately
after

\end{fulllineitems}

\index{finish\_run() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.finish_run}\pysiglinewithargsret{\bfcode{finish\_run}}{}{}
Operations to perform after the daemon thread stops running

\end{fulllineitems}

\index{list() (soprano.hpc.submitter.submit.Submitter static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.list}\pysiglinewithargsret{\strong{static }\bfcode{list}}{}{}
\end{fulllineitems}

\index{log() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.log}\pysiglinewithargsret{\bfcode{log}}{\emph{logtxt}}{}
\end{fulllineitems}

\index{next\_job() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.next_job}\pysiglinewithargsret{\bfcode{next\_job}}{}{}
Return a dictionary definition of the next job in line

\end{fulllineitems}

\index{set\_parameters() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.set_parameters}\pysiglinewithargsret{\bfcode{set\_parameters}}{}{}
Set additional parameters. In this generic example class it has
no arguments, but in specific implementations it will be used to
add more variables without overriding \_\_init\_\_.

\end{fulllineitems}

\index{setup\_job() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.setup_job}\pysiglinewithargsret{\bfcode{setup\_job}}{\emph{name}, \emph{args}, \emph{folder}}{}
Perform preparatory operations on the job

\end{fulllineitems}

\index{start() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.start}\pysiglinewithargsret{\bfcode{start}}{}{}
\end{fulllineitems}

\index{start\_run() (soprano.hpc.submitter.submit.Submitter method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.start_run}\pysiglinewithargsret{\bfcode{start\_run}}{}{}
Operations to perform when the daemon thread starts running

\end{fulllineitems}

\index{stop() (soprano.hpc.submitter.submit.Submitter static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.hpc.submitter.submit:soprano.hpc.submitter.submit.Submitter.stop}\pysiglinewithargsret{\strong{static }\bfcode{stop}}{\emph{fname}, \emph{subname}}{}
Stop Submitter process from filename and name,
return False if failed

\end{fulllineitems}


\end{fulllineitems}



\subsubsection{soprano.properties package}
\label{soprano.properties:module-soprano.properties}\label{soprano.properties:soprano-properties-package}\label{soprano.properties::doc}\index{soprano.properties (module)}
Contains classes, modules and functions relevant to Properties,
a catch-all term for things we might want to extract or calculate from
Atoms and AtomsCollections. Some will require running an external ASE
calculator first, some will just work on their own, some will require
some calculations and parameters.


\paragraph{Subpackages}
\label{soprano.properties:subpackages}

\subparagraph{soprano.properties.basic package}
\label{soprano.properties.basic:soprano-properties-basic-package}\label{soprano.properties.basic::doc}\label{soprano.properties.basic:module-soprano.properties.basic}\index{soprano.properties.basic (module)}
Module containing very basic AtomsProperties (the kind that only require
a couple of lines of code but are still pretty convenient to have at hand)


\subparagraph{Submodules}
\label{soprano.properties.basic:submodules}

\subparagraph{soprano.properties.basic.basic module}
\label{soprano.properties.basic.basic::doc}\label{soprano.properties.basic.basic:module-soprano.properties.basic.basic}\label{soprano.properties.basic.basic:soprano-properties-basic-basic-module}\index{soprano.properties.basic.basic (module)}
Implementation of some basic AtomsProperty classes
\index{CalcEnergy (class in soprano.properties.basic.basic)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.CalcEnergy}\pysiglinewithargsret{\strong{class }\code{soprano.properties.basic.basic.}\bfcode{CalcEnergy}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Property representing the energy calculated by an ASE calulator

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.basic.basic.CalcEnergy attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.CalcEnergy.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'calc\_energy'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.basic.basic.CalcEnergy attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.CalcEnergy.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.basic.basic.CalcEnergy static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.CalcEnergy.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{LatticeABC (class in soprano.properties.basic.basic)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeABC}\pysiglinewithargsret{\strong{class }\code{soprano.properties.basic.basic.}\bfcode{LatticeABC}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Property representing the axis-angles form of a structure's lattice

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] shape (tuple): the shape to give to the array
\item[] deg (bool): whether to give the angles in degrees instead of radians
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.basic.basic.LatticeABC attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeABC.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'lattice\_abc'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.basic.basic.LatticeABC attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeABC.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'shape': (2, 3), u'deg': False\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.basic.basic.LatticeABC static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeABC.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{shape}, \emph{deg}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{LatticeCart (class in soprano.properties.basic.basic)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeCart}\pysiglinewithargsret{\strong{class }\code{soprano.properties.basic.basic.}\bfcode{LatticeCart}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Property representing the Cartesian form of a structure's lattice

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] shape (tuple): the shape to give to the array
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.basic.basic.LatticeCart attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeCart.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'lattice\_cart'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.basic.basic.LatticeCart attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeCart.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'shape': (3, 3)\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.basic.basic.LatticeCart static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.LatticeCart.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{shape}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{NumAtoms (class in soprano.properties.basic.basic)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.NumAtoms}\pysiglinewithargsret{\strong{class }\code{soprano.properties.basic.basic.}\bfcode{NumAtoms}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Property representing the number of atoms in a structure

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.basic.basic.NumAtoms attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.NumAtoms.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'num\_atoms'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.basic.basic.NumAtoms attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.NumAtoms.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.basic.basic.NumAtoms static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.basic.basic:soprano.properties.basic.basic.NumAtoms.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.properties.castep package}
\label{soprano.properties.castep:module-soprano.properties.castep}\label{soprano.properties.castep::doc}\label{soprano.properties.castep:soprano-properties-castep-package}\index{soprano.properties.castep (module)}
Module containing AtomsProperties related specifically to CASTEP calculations.
Some of these can be looked up only in a CASTEP Calculator; others require
passing the path of the .castep file as a parameter and actually parsing its
contents.


\subparagraph{Submodules}
\label{soprano.properties.castep:submodules}

\subparagraph{soprano.properties.castep.castep module}
\label{soprano.properties.castep.castep:soprano-properties-castep-castep-module}\label{soprano.properties.castep.castep::doc}\label{soprano.properties.castep.castep:module-soprano.properties.castep.castep}\index{soprano.properties.castep.castep (module)}
Implementation of some CASTEP related AtomsProperties
\index{CastepEnthalpy (class in soprano.properties.castep.castep)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.castep.castep:soprano.properties.castep.castep.CastepEnthalpy}\pysiglinewithargsret{\strong{class }\code{soprano.properties.castep.castep.}\bfcode{CastepEnthalpy}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Enthalpy as found in the .castep file of a GeometryOptimization
calculation. If not present, this will fall back on the final free energy.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] castep\_path (str): the path in which the .castep file is to be found.
\item[] seedname\_info (str): the Atoms.info key that contains the seedname
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] of the .castep file. By default is `name'.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.castep.castep.CastepEnthalpy attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.castep.castep:soprano.properties.castep.castep.CastepEnthalpy.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'castep\_enthalpy'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.castep.castep.CastepEnthalpy attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.castep.castep:soprano.properties.castep.castep.CastepEnthalpy.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'seedname\_info': u'name', u'castep\_path': u'.'\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.castep.castep.CastepEnthalpy static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.castep.castep:soprano.properties.castep.castep.CastepEnthalpy.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{castep\_path}, \emph{seedname\_info}}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.properties.labeling package}
\label{soprano.properties.labeling:module-soprano.properties.labeling}\label{soprano.properties.labeling::doc}\label{soprano.properties.labeling:soprano-properties-labeling-package}\index{soprano.properties.labeling (module)}
Module containing AtomsProperties that relate to labeling a system's atoms,
molecules, hydrogen bonds and such based on their chemical properties.


\subparagraph{Submodules}
\label{soprano.properties.labeling:submodules}

\subparagraph{soprano.properties.labeling.labeling module}
\label{soprano.properties.labeling.labeling:soprano-properties-labeling-labeling-module}\label{soprano.properties.labeling.labeling:module-soprano.properties.labeling.labeling}\label{soprano.properties.labeling.labeling::doc}\index{soprano.properties.labeling.labeling (module)}
Implementation of AtomsProperties that relate to labeling of systems
\index{HydrogenBondTypes (class in soprano.properties.labeling.labeling)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.HydrogenBondTypes}\pysiglinewithargsret{\strong{class }\code{soprano.properties.labeling.labeling.}\bfcode{HydrogenBondTypes}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Assign MoleculeSites labels to atoms, then characterise existing hydrogen
bonds based on them, and return a list of such bonds detected in a system.
The bonds come in the form `\{0\}\textless{}\{1\},\{2\}\textgreater{}..\{3\}\textless{}\{4\}\textgreater{}', where \{0\} is the name
of the molecule containing the hydrogen, \{2\} is the hydrogen, \{1\} the atom
to which the hydrogen is bonded, \{3\} the name of the other molecule and
\{4\} the atom to which the hydrogen is hydrogen bonded.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\item[] save\_info (bool): if True, save the found hydrogen bond types as part
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] of the Atoms object info. By default True.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] hydrogen\_bond\_types (list): A list containing info characterising the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] hydrogen bonds present in the system in a
\item[] detailed way.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.labeling.labeling.HydrogenBondTypes attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.HydrogenBondTypes.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'hydrogen\_bond\_types'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.labeling.labeling.HydrogenBondTypes attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.HydrogenBondTypes.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False, u'save\_info': True\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.labeling.labeling.HydrogenBondTypes static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.HydrogenBondTypes.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}, \emph{save\_info}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{MoleculeSites (class in soprano.properties.labeling.labeling)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.MoleculeSites}\pysiglinewithargsret{\strong{class }\code{soprano.properties.labeling.labeling.}\bfcode{MoleculeSites}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Assigns univoque labels to atoms belonging to molecules by exploiting
network topology. Atoms can have the same label, but only if they're
fundamentally indistinguishable in the molecule's chemical context
(for example, three hydrogen atoms on a CH3 group). The molecule will be
described by a characteristic string and by a series of labels in the
format {[}element{]}\_{[}number{]}. These sites will be saved by default and can
be used for better insight when carrying out other analysis.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\item[] save\_info (bool): if True, save the found molecular sites as part of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the Atoms object info. By default True.
\end{DUlineblock}
\item[] save\_asarray (bool): if True the molecular site names are also saved
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] as an array of the molecule selection.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecular\_sites (dict): A dictionary containing info characterising
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the molecule's chemical sites unequivocally.
\item[] These are a string representation of the
\item[] molecule itself and a dictionary linking
\item[] atomic indices (as found in the molecule in
\item[] AtomSelection form) to site labels.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.labeling.labeling.MoleculeSites attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.MoleculeSites.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecule\_sites'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.labeling.labeling.MoleculeSites attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.MoleculeSites.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False, u'save\_info': True, u'save\_asarray': False\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.labeling.labeling.MoleculeSites static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.labeling.labeling:soprano.properties.labeling.labeling.MoleculeSites.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}, \emph{save\_info}, \emph{save\_asarray}}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.properties.linkage package}
\label{soprano.properties.linkage::doc}\label{soprano.properties.linkage:soprano-properties-linkage-package}\label{soprano.properties.linkage:module-soprano.properties.linkage}\index{soprano.properties.linkage (module)}
Module containing AtomsProperties that relate to linkage properties of a given
system, self-correlation etc.


\subparagraph{Submodules}
\label{soprano.properties.linkage:submodules}

\subparagraph{soprano.properties.linkage.linkage module}
\label{soprano.properties.linkage.linkage::doc}\label{soprano.properties.linkage.linkage:module-soprano.properties.linkage.linkage}\label{soprano.properties.linkage.linkage:soprano-properties-linkage-linkage-module}\index{soprano.properties.linkage.linkage (module)}
Implementation of AtomsProperties that relate to linkage of atoms
\index{Bonds (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Bonds}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{Bonds}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Produces an array of tuples identifying all bonds existing within the 
system (calculated using Van der Waals radii). The tuples are structured
as:

(atom\_1, atom\_2, atom\_2\_cell, bond\_length)

with atom\_1 and atom\_2 being indices and atom\_2\_cell being an array of
integers identifying the unit cell to which atom\_2 belongs with respect
to atom\_1 (which is assumed to be in (0,0,0), the central cell). This is
to account for the possibility of course that the bond exists through the
periodic boundary. WARNING: the possibility of an atom bonding with
another throughout two different periodic boundaries is not accounted for.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vdw\_set(\{ase, jmol\}): set of Van der Waals radii to use. Default is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the one extracted from JMol.
\end{DUlineblock}
\item[] vdw\_scale (float): scaling factor to apply to the base Van der Waals
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] radii values. Values bigger than one make for more
\item[] tolerant bonds.
\end{DUlineblock}
\item[] default\_vdw (float): default Van der Waals radius for species for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] whom no data is available.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] bonds({[}tuple{]}): list of bonds in the form of 3-tuples structured as
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] explained above
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.Bonds attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Bonds.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'bonds'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.Bonds attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Bonds.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'default\_vdw': 2.0, u'vdw\_scale': 1.0, u'vdw\_set': u'jmol'\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.Bonds static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Bonds.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{vdw\_set}, \emph{vdw\_scale}, \emph{default\_vdw}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{CoordinationHistogram (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.CoordinationHistogram}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{CoordinationHistogram}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Produces an histogram representing, for each pair of species present in
the system, how many atoms of species 1 have n bonds with species 2, n
being the histogram bins. The histogram is topped at a `maximum
coordination' parameter which is 6 by default but can be user defined;
the last bin represents all higher values (so by default `6 or more').
Two species or lists of species can be given if one wants to restrict the
search; otherwise a full histogram for all pairs of species is returned.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vdw\_set(\{ase, jmol\}): set of Van der Waals radii to use. Default is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the one extracted from JMol.
\end{DUlineblock}
\item[] vdw\_scale (float): scaling factor to apply to the base Van der Waals
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] radii values. Values bigger than one make for more
\item[] tolerant bonds.
\end{DUlineblock}
\item[] default\_vdw (float): default Van der Waals radius for species for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] whom no data is available.
\end{DUlineblock}
\item[] species\_1 (str or {[}str{]}): list of species to compute the histogram
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] for. By default all of them.
\end{DUlineblock}
\item[] species\_2 (str or {[}str{]}): list of species whose coordination with 
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] species\_1 should be checked. By default all
\item[] of them.
\end{DUlineblock}
\item[] max\_coord (int): what should be the largest coordination number
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] considered for an atom (default 6).
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] coord\_hist (dict): dictionary of dictionaries indexed by species\_1
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] followed by species\_2. The elements are arrays of
\item[] integers constituting the histogram.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.CoordinationHistogram attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.CoordinationHistogram.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'coord\_histogram'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.CoordinationHistogram attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.CoordinationHistogram.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'species\_2': None, u'species\_1': None, u'default\_vdw': 2.0, u'max\_coord': 6, u'vdw\_set': u'jmol', u'vdw\_scale': 1.0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.CoordinationHistogram static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.CoordinationHistogram.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{vdw\_set}, \emph{vdw\_scale}, \emph{default\_vdw}, \emph{species\_1}, \emph{species\_2}, \emph{max\_coord}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{HydrogenBonds (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBonds}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{HydrogenBonds}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Hydrogen Bonds

Produces a dictionary containing the atom indices defining hydrogen bonds
detected in the system - if required, classified by type. By default only
O and N atoms are considered for hydrogen bonds (OH..O, OH..N and so on).
The type is defined as AH..B where A is the symbol of the atom directly
bonded to the proton and B the one of the hydrogen bonded one.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vdw\_set(\{ase, jmol\}): set of Van der Waals radii to use. Default is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the one extracted from JMol.
\end{DUlineblock}
\item[] vdw\_scale (float): scaling factor to apply to the base Van der Waals
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] radii values. Values bigger than one make for more
\item[] tolerant molecules.
\end{DUlineblock}
\item[] default\_vdw (float): default Van der Waals radius for species for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] whom no data is available.
\end{DUlineblock}
\item[] hbond\_elems ({[}str{]}): chemical symbols of elements considered capable
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] of forming hydrogen bonds (by default O and N)
\end{DUlineblock}
\item[] max\_length (float): maximum A-B length of the hydrogen bond in
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Angstrom - default is 3.5 Ang
\end{DUlineblock}
\item[] max\_angle (float): maximum A-H/A-B angle in the hydrogen bond in
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] degrees - default is 45 deg
\end{DUlineblock}
\item[] save\_info (bool): if True, save the found hydrogen bonds as part of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the Atoms object info. By default True.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] hbondss ({[}dict{]}): list of hydrogen bonds detected
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] in the system by type (can contain empty arrays).
\item[] For each hydrogen bond we give index of the H
\item[] atom, index and unit cell of the A atom (the one
\item[] directly bonded), index and unit cell of the B atom
\item[] (the one that's hydrogen bonded), length and angle
\item[] in degrees.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.HydrogenBonds attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBonds.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'hydrogen\_bonds'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.HydrogenBonds attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBonds.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'save\_info': True, u'hbond\_elems': {[}u'O', u'N'{]}, u'default\_vdw': 2.0, u'max\_length': 3.5, u'vdw\_set': u'jmol', u'vdw\_scale': 1.0, u'max\_angle': 45.0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.HydrogenBonds static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBonds.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{vdw\_set}, \emph{vdw\_scale}, \emph{default\_vdw}, \emph{hbond\_elems}, \emph{max\_length}, \emph{max\_angle}, \emph{save\_info}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{HydrogenBondsNumber (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBondsNumber}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{HydrogenBondsNumber}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Number of hydrogen bonds detected in this system, classified by type.
By default will use already existing hydrogen bonds if they're present as
a saved array in the system.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the hydrogen bonds
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] even if already present.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] hbonds\_n (int): number of hydrogen bonds found
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.HydrogenBondsNumber attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBondsNumber.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'hydrogen\_bonds\_n'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.HydrogenBondsNumber attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBondsNumber.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.HydrogenBondsNumber static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.HydrogenBondsNumber.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{LinkageList (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.LinkageList}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{LinkageList}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Produces an array containing the atomic pair distances in a system,
reduced to their shortest periodic version and sorted min to max.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] size (int): maximum number of distances to include. If not present,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] all of them will be included. If present, arrays will be
\item[] cut or padded to reach this sizeber.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] link\_list ({[}float{]}): sorted list of interatomic linkage distances
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.LinkageList attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.LinkageList.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'linkage\_list'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.LinkageList attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.LinkageList.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'size': 0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.LinkageList static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.LinkageList.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{size}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{MoleculeCOMLinkage (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeCOMLinkage}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{MoleculeCOMLinkage}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Linkage list - following the same criteria as the atomic one - calculated
for the centers of mass of the molecules present in the system. By default
will use already existing molecules if they're present as a saved array in
the system.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\item[] size (int): maximum number of distances to include. If not present,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] all of them will be included. If present, arrays will be
\item[] cut or padded to reach this sizeber.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecule\_linkage ({[}float{]}): distances between all centers of mass of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecules in the system, sorted.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.MoleculeCOMLinkage attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeCOMLinkage.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecule\_com\_linkage'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.MoleculeCOMLinkage attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeCOMLinkage.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False, u'size': 0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.MoleculeCOMLinkage static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeCOMLinkage.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}, \emph{size}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{MoleculeMass (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeMass}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{MoleculeMass}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Total mass of each of the molecules detected in this system. By default
will use already existing molecules if they're present as a saved array in
the system.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\item[] size (int): maximum number of distances to include. If not present,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] all of them will be included. If present, arrays will be
\item[] cut or padded to reach this sizeber.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecule\_m ({[}float{]}): mass of each of the molecules present, sorted.
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.MoleculeMass attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeMass.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecule\_mass'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.MoleculeMass attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeMass.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False, u'size': 0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.MoleculeMass static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeMass.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}, \emph{size}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{MoleculeNumber (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeNumber}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{MoleculeNumber}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Number of molecules detected in this system. By default will use already
existing molecules if they're present as a saved array in the system.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecule\_n (int): number of molecules found
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.MoleculeNumber attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeNumber.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecule\_n'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.MoleculeNumber attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeNumber.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.MoleculeNumber static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeNumber.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{MoleculeRelativeRotation (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeRelativeRotation}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{MoleculeRelativeRotation}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

A list of relative rotations between molecules. Uses the inertia tensor
eigenvectors to establish a local frame for each molecule, then uses
quaternions to define a rotational distance between molecules. It then
produces a list of geodesic distances between these quaternions.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] force\_recalc (bool): if True, always recalculate the molecules even if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] already present.
\end{DUlineblock}
\item[] size (int): maximum number of distances to include. If not present,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] all of them will be included. If present, arrays will be
\item[] cut or padded to reach this size.
\end{DUlineblock}
\item[] twist\_axis ({[}float{]}): if present, only compare the Twist component of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] quaternion along the given axis. The Twist/Swing
\item[] decomposition splits a quaternion in a rotation
\item[] around an axis and one around an orthogonal
\item[] direction. Only one between this and swing\_plane
\item[] can be present.
\end{DUlineblock}
\item[] swing\_plane ({[}float{]}): if present, only compare the Swing component of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] quaternion along the given axis. The Twist/Swing
\item[] decomposition splits a quaternion in a rotation
\item[] around an axis and one around an orthogonal
\item[] direction. Only one between this and twist\_axis
\item[] can be present.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecule\_relrot ({[}float{]}): list of relative rotations, as quaternion
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] distances, with the required ordering.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.MoleculeRelativeRotation attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeRelativeRotation.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecule\_rel\_rotation'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.MoleculeRelativeRotation attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeRelativeRotation.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'force\_recalc': False, u'twist\_axis': None, u'swing\_plane': None, u'size': 0\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.MoleculeRelativeRotation static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.MoleculeRelativeRotation.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{force\_recalc}, \emph{size}, \emph{swing\_plane}, \emph{twist\_axis}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{Molecules (class in soprano.properties.linkage.linkage)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Molecules}\pysiglinewithargsret{\strong{class }\code{soprano.properties.linkage.linkage.}\bfcode{Molecules}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Produces an array containing multiple AtomSelection objects representing
molecules in the system as found by connecting atoms closer than the half
sum of their Van der Waals radii. It will return the entire unit cell if
the system can not be split in molecules at all.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vdw\_set(\{ase, jmol\}): set of Van der Waals radii to use. Default is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the one extracted from JMol.
\end{DUlineblock}
\item[] vdw\_scale (float): scaling factor to apply to the base Van der Waals
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] radii values. Values bigger than one make for more
\item[] tolerant molecules.
\end{DUlineblock}
\item[] default\_vdw (float): default Van der Waals radius for species for
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] whom no data is available.
\end{DUlineblock}
\item[] save\_info (bool): if True, save the found molecules as part of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] Atoms object info. By default True.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] molecules ({[}AtomSelection{]}): list of molecules in the form of
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] AtomSelection objects.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.linkage.linkage.Molecules attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Molecules.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'molecules'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.linkage.linkage.Molecules attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Molecules.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'default\_vdw': 2.0, u'vdw\_scale': 1.0, u'save\_info': True, u'vdw\_set': u'jmol'\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.linkage.linkage.Molecules static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.linkage.linkage:soprano.properties.linkage.linkage.Molecules.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{vdw\_set}, \emph{vdw\_scale}, \emph{default\_vdw}, \emph{save\_info}}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.properties.symmetry package}
\label{soprano.properties.symmetry::doc}\label{soprano.properties.symmetry:soprano-properties-symmetry-package}\label{soprano.properties.symmetry:module-soprano.properties.symmetry}\index{soprano.properties.symmetry (module)}
Module containing AtomProperties that pertain to symmetry detection.
Depends on having the Python bindings to SPGLIB installed on the system.


\subparagraph{Submodules}
\label{soprano.properties.symmetry:submodules}

\subparagraph{soprano.properties.symmetry.symmetry module}
\label{soprano.properties.symmetry.symmetry:module-soprano.properties.symmetry.symmetry}\label{soprano.properties.symmetry.symmetry::doc}\label{soprano.properties.symmetry.symmetry:soprano-properties-symmetry-symmetry-module}\index{soprano.properties.symmetry.symmetry (module)}
Implementation of AtomProperties that relate to symmetry
\index{SymmetryDataset (class in soprano.properties.symmetry.symmetry)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.symmetry.symmetry:soprano.properties.symmetry.symmetry.SymmetryDataset}\pysiglinewithargsret{\strong{class }\code{soprano.properties.symmetry.symmetry.}\bfcode{SymmetryDataset}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Extracts SPGLIB's standard symmetry dataset from a given system, including
spacegroup symbol, symmetry operations etc.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] symprec (float): distance tolerance, in Angstroms, applied when
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] searching symmetry.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] symm\_dataset (dict): dictionary of symmetry information
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.symmetry.symmetry.SymmetryDataset attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.symmetry.symmetry:soprano.properties.symmetry.symmetry.SymmetryDataset.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'symmetry\_dataset'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.symmetry.symmetry.SymmetryDataset attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.symmetry.symmetry:soprano.properties.symmetry.symmetry.SymmetryDataset.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'symprec': 1e-05\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.symmetry.symmetry.SymmetryDataset static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.symmetry.symmetry:soprano.properties.symmetry.symmetry.SymmetryDataset.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{symprec}}{}
\end{fulllineitems}


\end{fulllineitems}



\subparagraph{soprano.properties.transform package}
\label{soprano.properties.transform:soprano-properties-transform-package}\label{soprano.properties.transform::doc}\label{soprano.properties.transform:module-soprano.properties.transform}\index{soprano.properties.transform (module)}
Module containing  a special set of AtomsProperties that transform an Atoms
object into another (by translating, rotating or mirroring all or some ions,
and so on). These all accept an Atoms object and some parameters and return
an Atoms object as well. Default behaviour for the .get method in most cases
will be to do nothing at all, these properties are meant to be instantiated.


\subparagraph{Submodules}
\label{soprano.properties.transform:submodules}

\subparagraph{soprano.properties.transform.transform module}
\label{soprano.properties.transform.transform:soprano-properties-transform-transform-module}\label{soprano.properties.transform.transform::doc}\label{soprano.properties.transform.transform:module-soprano.properties.transform.transform}\index{soprano.properties.transform.transform (module)}
Implementation of AtomsProperties that transform the instance in some
way
\index{Mirror (class in soprano.properties.transform.transform)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Mirror}\pysiglinewithargsret{\strong{class }\code{soprano.properties.transform.transform.}\bfcode{Mirror}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Returns an Atoms object with some or all the atoms reflected with either a
given center or a given plane. Absolute or scaled coordinates may be used.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection): selection object defining which atoms to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] act on. By default, all of them.
\end{DUlineblock}
\item[] center ({[}float{]}*3): center around which the reflection should take
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] place. By default the origin of the axes. Can't be
\item[] present at the same time as plane.
\end{DUlineblock}
\item[] plane ({[}float{]}*4): plane with respect to which the reflection should
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] take place, in the form {[}a, b, c, d{]} parameters of
\item[] the plane equation.
\item[] By default is not used. Can't be present at the
\item[] same time as center.
\end{DUlineblock}
\item[] scaled (bool): if True, treat the input vector as expressed in scaled,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] not absolute, coordinates.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] reflected (ase.Atoms): Atoms object with the reflection performed.
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.transform.transform.Mirror attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Mirror.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'reflected'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.transform.transform.Mirror attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Mirror.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'scaled': False, u'selection': None, u'center': None, u'plane': None\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.transform.transform.Mirror static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Mirror.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{selection}, \emph{**kwargs}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{Regularise (class in soprano.properties.transform.transform)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Regularise}\pysiglinewithargsret{\strong{class }\code{soprano.properties.transform.transform.}\bfcode{Regularise}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Regularize

Perform a translation by a vector calculated to cancel out the effect of
global translational symmetry. In theory, given two copies of the same
system that only differ by a translation of all atoms in the unit cell,
this should produce two systems that overlap perfectly. Can be used to
compare slightly different systems if they're similar enough. If a
selection is given, only those atoms will be used to calculate the center,
but the translation will still be applied to all atoms. The same atoms
have to be used in all systems for comparisons to make sense (for example
one might use all the heavy atoms and not include hydrogens).

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection): selection object defining which atoms to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] act on. By default, all of them.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] regularised (ase.Atoms): Atoms object translated by the regularizing
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vector.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.transform.transform.Regularise attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Regularise.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'regularised'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.transform.transform.Regularise attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Regularise.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'selection': None\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.transform.transform.Regularise static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Regularise.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{selection}, \emph{**kwargs}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{Rotate (class in soprano.properties.transform.transform)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Rotate}\pysiglinewithargsret{\strong{class }\code{soprano.properties.transform.transform.}\bfcode{Rotate}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Returns an Atoms object with some or all the atoms rotated by a given
quaternion and with a given center. Absolute or scaled coordinates may be
used.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection): selection object defining which atoms to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] act on. By default, all of them.
\end{DUlineblock}
\item[] center ({[}float{]}*3): center around which the rotation should take
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] place. By default the origin of the axes.
\end{DUlineblock}
\item[] quaternion (ase.quaternions.Quaternion): quaternion expressing the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] rotation that should be
\item[] applied.
\end{DUlineblock}
\item[] scaled (bool): if True, treat the input vector as expressed in scaled,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] not absolute, coordinates.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] rotated (ase.Atoms): Atoms object with the rotation performed.
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.transform.transform.Rotate attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Rotate.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'rotated'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.transform.transform.Rotate attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Rotate.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'scaled': False, u'quaternion': None, u'selection': None, u'center': {[}0, 0, 0{]}\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.transform.transform.Rotate static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Rotate.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{selection}, \emph{**kwargs}}{}
\end{fulllineitems}


\end{fulllineitems}

\index{Translate (class in soprano.properties.transform.transform)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Translate}\pysiglinewithargsret{\strong{class }\code{soprano.properties.transform.transform.}\bfcode{Translate}}{\emph{name=None}, \emph{**params}}{}
Bases: {\hyperref[soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty]{\emph{\code{soprano.properties.atomsproperty.AtomsProperty}}}}

Returns an Atoms object with some or all the atoms translated by a given
vector. Absolute or scaled coordinates may be used.

\begin{DUlineblock}{0em}
\item[] Parameters:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection): selection object defining which atoms to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] act on. By default, all of them.
\end{DUlineblock}
\item[] vector ({[}float{]}*3): vector by which to translate the atoms.
\item[] scaled (bool): if True, treat the input vector as expressed in scaled,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] not absolute, coordinates.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] translated (ase.Atoms): Atoms object with the translation performed.
\end{DUlineblock}
\end{DUlineblock}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.transform.transform.Translate attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Translate.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'translated'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.transform.transform.Translate attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Translate.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{u'vector': {[}0, 0, 0{]}, u'scaled': False, u'selection': None\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.transform.transform.Translate static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.transform.transform:soprano.properties.transform.transform.Translate.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{selection}, \emph{**kwargs}}{}
\end{fulllineitems}


\end{fulllineitems}



\paragraph{Submodules}
\label{soprano.properties:submodules}

\subparagraph{soprano.properties.atomsproperty module}
\label{soprano.properties.atomsproperty:module-soprano.properties.atomsproperty}\label{soprano.properties.atomsproperty:soprano-properties-atomsproperty-module}\label{soprano.properties.atomsproperty::doc}\index{soprano.properties.atomsproperty (module)}
Definition of AtomsProperty class.

A generic template class that specific Properties will inherit from.
\index{AtomsProperty (class in soprano.properties.atomsproperty)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty}\pysiglinewithargsret{\strong{class }\code{soprano.properties.atomsproperty.}\bfcode{AtomsProperty}}{\emph{name=None}, \emph{**params}}{}
Bases: \code{object}

Initialize an AtomsProperty and set its parameters.
The AtomsProperty instance can then be called with a structure as its
only argument to get the property with the given parameters.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): a name to give to this specific instance of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property (will be used to store it as array if
\item[] requested)
\end{DUlineblock}
\item[] params: named arguments specific to this type of property
\end{DUlineblock}
\end{DUlineblock}
\index{default\_name (soprano.properties.atomsproperty.AtomsProperty attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty.default_name}\pysigline{\bfcode{default\_name}\strong{ = u'generic\_property'}}
\end{fulllineitems}

\index{default\_params (soprano.properties.atomsproperty.AtomsProperty attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty.default_params}\pysigline{\bfcode{default\_params}\strong{ = \{\}}}
\end{fulllineitems}

\index{extract() (soprano.properties.atomsproperty.AtomsProperty static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty.extract}\pysiglinewithargsret{\strong{static }\bfcode{extract}}{\emph{s}, \emph{**params}}{}
Extract the given property with given parameters from an Atoms
object.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] s (ase.Atoms): the structure from which to extract the property
\item[] params: named arguments specific to this type of property
\item[] 
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property: the value of the property for the given structure and
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] parameters
\item[] 
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get() (soprano.properties.atomsproperty.AtomsProperty class method)}

\begin{fulllineitems}
\phantomsection\label{soprano.properties.atomsproperty:soprano.properties.atomsproperty.AtomsProperty.get}\pysiglinewithargsret{\strong{classmethod }\bfcode{get}}{\emph{s}, \emph{store\_array=False}}{}
Extract the given property using the default parameters
on an Atoms object s

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] s (ase.Atoms or AtomsCollection): the structure or collection
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] from which to extract the
\item[] property
\end{DUlineblock}
\item[] store\_array (bool): if s is a collection, whether to store the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] resulting data as an array in the collection
\item[] using the default name for this property
\item[] 
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] property: the value of the property for the given structure or
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] a list of values if a collection has been passed
\item[] 
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}


\end{fulllineitems}



\subsection{Submodules}
\label{soprano:submodules}

\subsubsection{soprano.selection module}
\label{soprano.selection:module-soprano.selection}\label{soprano.selection::doc}\label{soprano.selection:soprano-selection-module}\index{soprano.selection (module)}
selection.py

Contains the definition of an AtomSelection class,
namely a group of selected atoms for a given structure,
and methods to build it.
\index{AtomSelection (class in soprano.selection)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection}\pysiglinewithargsret{\strong{class }\code{soprano.selection.}\bfcode{AtomSelection}}{\emph{atoms}, \emph{sel\_indices}, \emph{authenticate=True}}{}
Bases: \code{object}

AtomSelection object.

An AtomSelection represents a group of atoms from an ASE Atoms object.
It keeps track of them and can be used to perform operations on them
(for example geometrical transformation or extraction of specific
properties).
It does not keep track of the original Atoms object it's been created
from, but can be ``authenticated'' to verify that it is indeed operating
consistently on the same structure. It also provides a series of static
methods to build selections with various criteria.

Initialize the AtomSelection.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (ase.Atoms): the atoms object on which the selection is
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] applied
\end{DUlineblock}
\item[] sel\_indices (list{[}int{]}): the list of indices of the atoms that
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] are to be selected
\end{DUlineblock}
\item[] authenticate (Optional{[}bool{]}): whether to use hashing to confirm
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] the identity of the atoms object
\item[] we're operating with
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}
\index{all() (soprano.selection.AtomSelection static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.all}\pysiglinewithargsret{\strong{static }\bfcode{all}}{\emph{atoms}}{}
Generate a selection for the given Atoms object of all atoms.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (ase.Atoms): Atoms object on which to perform selection
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection)
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{from\_box() (soprano.selection.AtomSelection static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.from_box}\pysiglinewithargsret{\strong{static }\bfcode{from\_box}}{\emph{atoms}, \emph{abc0}, \emph{abc1}, \emph{periodic=False}, \emph{scaled=False}}{}
Generate a selection for the given Atoms object of all atoms within
a given box volume.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (ase.Atoms): Atoms object on which to perform selection
\item[] abc0 ({[}float, float, float{]}): bottom corner of box
\item[] abc1 ({[}float, float, float{]}): top corner of box
\item[] periodic (Optional{[}bool{]}): if True, include periodic copies of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms
\end{DUlineblock}
\item[] scaled (Optional{[}bool{]}): if True, consider scaled (fractional)
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] coordinates instead of absolute ones
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection)
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{from\_element() (soprano.selection.AtomSelection static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.from_element}\pysiglinewithargsret{\strong{static }\bfcode{from\_element}}{\emph{atoms}, \emph{element}}{}
Generate a selection for the given Atoms object of all atoms of a
specific element.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (ase.Atoms): Atoms object on which to perform selection
\item[] element (str): symbol of the element to select
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection)
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{from\_sphere() (soprano.selection.AtomSelection static method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.from_sphere}\pysiglinewithargsret{\strong{static }\bfcode{from\_sphere}}{\emph{atoms}, \emph{center}, \emph{r}, \emph{periodic=False}, \emph{scaled=False}}{}
Generate a selection for the given Atoms object of all atoms within
a given spherical volume.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms (ase.Atoms): Atoms object on which to perform selection
\item[] center ({[}float, float, float{]}): center of the sphere
\item[] r (float): radius of the sphere
\item[] periodic (Optional{[}bool{]}): if True, include periodic copies of the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] atoms
\end{DUlineblock}
\item[] scaled (Optional{[}bool{]}): if True, consider scaled (fractional)
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] coordinates instead of absolute ones
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] selection (AtomSelection)
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{get\_array() (soprano.selection.AtomSelection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.get_array}\pysiglinewithargsret{\bfcode{get\_array}}{\emph{name}}{}
Retrieve a previously stored data array.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the array to be set or created
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] array (np.ndarray): array of data to be saved
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{indices (soprano.selection.AtomSelection attribute)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.indices}\pysigline{\bfcode{indices}}
\end{fulllineitems}

\index{set\_array() (soprano.selection.AtomSelection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.set_array}\pysiglinewithargsret{\bfcode{set\_array}}{\emph{name}, \emph{array}}{}
Save an array of given name containing arbitraty information
tied to the selected atoms.
This must match the length of the selection and will be passed on to
any Atoms objects created with .subset.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] name (str): name of the array to be set or created
\item[] array (np.ndarray): array of data to be saved
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{subset() (soprano.selection.AtomSelection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.subset}\pysiglinewithargsret{\bfcode{subset}}{\emph{atoms}}{}
Generate an Atoms object containing only the selected atoms.

\end{fulllineitems}

\index{validate() (soprano.selection.AtomSelection method)}

\begin{fulllineitems}
\phantomsection\label{soprano.selection:soprano.selection.AtomSelection.validate}\pysiglinewithargsret{\bfcode{validate}}{\emph{atoms}}{}
Check that the given Atoms object validates with this selection.

\end{fulllineitems}


\end{fulllineitems}



\subsubsection{soprano.utils module}
\label{soprano.utils:soprano-utils-module}\label{soprano.utils:module-soprano.utils}\label{soprano.utils::doc}\index{soprano.utils (module)}
utils.py

Contains package-wide useful routines that don't fall under any specific
category. Many of these handle common operations involving periodicity,
conversions between different representations etc.
\index{abc2cart() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.abc2cart}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{abc2cart}}{\emph{abc}}{}
Transforms an axes and angles representation of lattice parameters
into a Cartesian one

\end{fulllineitems}

\index{cart2abc() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.cart2abc}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{cart2abc}}{\emph{cart}}{}
Transforms a Cartesian representation of lattice parameters
into an axes and angles one

\end{fulllineitems}

\index{hkl2d2\_matgen() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.hkl2d2_matgen}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{hkl2d2\_matgen}}{\emph{abc}}{}
Generate a matrix that turns hkl indices into inverse crystal
plane distances for a given lattice in ABC form

\end{fulllineitems}

\index{inspect\_args() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.inspect_args}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{inspect\_args}}{\emph{f}}{}
\end{fulllineitems}

\index{inv\_plane\_dist() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.inv_plane_dist}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{inv\_plane\_dist}}{\emph{hkl}, \emph{hkl2d2}}{}
Calculate inverse planar distance for a given set of
Miller indices h, k, l.

\end{fulllineitems}

\index{is\_string() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.is_string}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{is\_string}}{\emph{s}}{}
Checks whether s is a string, with Python 2 and 3 compatibility

\end{fulllineitems}

\index{list\_distance() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.list_distance}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{list\_distance}}{\emph{l1}, \emph{l2}}{}
Return an integer distance between two lists (number of differing
elements)

\end{fulllineitems}

\index{minimum\_periodic() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.minimum_periodic}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{minimum\_periodic}}{\emph{v}, \emph{latt\_cart}}{}
Find the shortest periodic equivalent vector for a list of vectors and a
given lattice.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] v (np.ndarray): list of 3-vectors representing points or vectors to
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] reduce to their closest periodic version
\end{DUlineblock}
\item[] latt\_cart (np.ndarray): unit cell in cartesian form
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] v\_period (np.ndarray): array with the same shape as v, containing the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] vectors in periodic reduced form
\end{DUlineblock}
\item[] v\_cells (np.ndarray): array of triples of ints, corresponding to the
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] cells from which the various periodic copies of
\item[] the vectors were taken. For an unchanged vector
\item[] will be all {[}0,0,0{]}
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{minimum\_supcell() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.minimum_supcell}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{minimum\_supcell}}{\emph{max\_r, latt\_cart=None, r\_matrix=None, pbc={[}True, True, True{]}}}{}
Generate the bounds for a supercell containing a sphere
of given radius, knowing the unit cell.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] max\_r (float): radius of the sphere contained in the supercell
\item[] latt\_cart (np.ndarray): unit cell in cartesian form
\item[] r\_matrix (np.ndarray): matrix for the quadratic form returning
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] r\textasciicircum{}2 for this supercell.
\item[] Alternative to latt\_cart, for a direct
\item[] space cell would be equal to
\item[] np.dot(latt\_cart, latt\_cart.T)
\end{DUlineblock}
\item[] pbc ({[}bool, bool, bool{]}): periodic boundary conditions - if
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] a boundary is not periodic the
\item[] range returned will always be zero
\item[] in that dimension
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] shape (tuple{[}int{]}):  shape of the supercell to be built.
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{periodic\_center() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.periodic_center}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{periodic\_center}}{\emph{v\_frac}}{}
Alright, how does this work?
Basically, we're looking for the point, inside the unit cell, which
minimizes the sum of the squared distance from all ions.
Of course we need to consider the periodic boundaries. So the distance
on a single axis isn't simply abs(x), but a triangular wave.
Fun times! A triangular wave can be represented as a Fourier series. And
we can truncate that series to the first term because the minimum
basically stays the same and get:
sum((x-x\_i)**2) \textasciitilde{} sum((4/pi**2*sin(2*pi*(x-x\_i-1/4))+0.5)**2)
All the factors depend on the fact that we need to move the triangular
wave to the interval {[}0,1{]} and center it so that it's 0 for x-x\_i == 0.
It gets better! We take the derivative of this thing and look for a spot
where it becomes zero. The derivative is kind of a trigonometric
monstrosity but we can solve the equation by setting t = e\textasciicircum{}(2*pi*1.0j*x)
and then replacing cosines and sines with it. As a result, we get an
equation of 4th degree in t. And then we solve that with numpy.roots, take
the phase, turn that into a coordinate, find the one corresponding to the
absolute minimum.
All of which we can perform independently on each of the three axes
because the function is just the sum of the three components:
sum((r-r\_i)**2) = sum((x-x\_i)**2) + sum((y-y\_i)**2) + sum((z-z\_i)**2).
And there you go! Problem solved.

\end{fulllineitems}

\index{progbar() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.progbar}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{progbar}}{\emph{i}, \emph{i\_max}, \emph{bar\_len=20}, \emph{spinner=True}, \emph{spin\_rate=3.0}}{}
A textual progress bar for the command line

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] i (int): current progress index
\item[] max\_i (int): final progress index
\item[] bar\_len (Optional{[}int{]}): length in characters of the bar (no brackets)
\item[] spinner (Optional{[}bool{]}): show a spinner at the end
\item[] spin\_rate (Optional{[}float{]}): spinner rotation speed (turns per full
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] progress)
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] bar (str): a progress bar formatted as requested
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{replace\_folder() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.replace_folder}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{replace\_folder}}{\emph{path}, \emph{new\_folder}}{}
Replace the folder of the given path with a new one

\end{fulllineitems}

\index{seedname() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.seedname}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{seedname}}{\emph{path}}{}
Get the filename (with no extension) from a full path

\end{fulllineitems}

\index{supcell\_gridgen() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.supcell_gridgen}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{supcell\_gridgen}}{\emph{latt\_cart}, \emph{shape}}{}
Generate a full linearized grid for a supercell with r\_bounds
and a base unit cell in Cartesian form.

\begin{DUlineblock}{0em}
\item[] Args:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] latt\_cart (np.ndarray): unit cell in cartesian form
\item[] shape (tuple{[}int{]}):  shape of the supercell to be built,
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] as returned by minimum\_supcell.
\end{DUlineblock}
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Returns:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] neigh\_i\_grid (np.ndarray): supercell grid in fractional coordinates
\item[] neigh\_grid (np.ndarray): supercell grid in cartesian coordinates
\end{DUlineblock}
\end{DUlineblock}

\begin{DUlineblock}{0em}
\item[] Raises:
\item[]
\begin{DUlineblock}{\DUlineblockindent}
\item[] ValueError: if some of the arguments are invalid
\end{DUlineblock}
\end{DUlineblock}

\end{fulllineitems}

\index{swing\_twist\_decomp() (in module soprano.utils)}

\begin{fulllineitems}
\phantomsection\label{soprano.utils:soprano.utils.swing_twist_decomp}\pysiglinewithargsret{\code{soprano.utils.}\bfcode{swing\_twist\_decomp}}{\emph{quat}, \emph{axis}}{}
Perform a Swing*Twist decomposition of a Quaternion. This splits the
quaternion in two: one containing the rotation around axis (Twist), the
other containing the rotation around a vector parallel to axis (Swing).

Returns two quaternions: Swing, Twist.

\end{fulllineitems}

\phantomsection\label{index:module-soprano}\index{soprano (module)}
Soprano

A Python library to crack crystals
by Simone Sturniolo

Copyright (C) 2016 - Science and Technology Facility Council

Soprano is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Soprano is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see \textless{}\href{http://www.gnu.org/licenses/}{http://www.gnu.org/licenses/}\textgreater{}.


\chapter{Indices and tables}
\label{index:indices-and-tables}\begin{itemize}
\item {} 
\DUspan{xref,std,std-ref}{genindex}

\item {} 
\DUspan{xref,std,std-ref}{modindex}

\item {} 
\DUspan{xref,std,std-ref}{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{s}
\item {\texttt{soprano}}, \pageref{soprano:module-soprano}
\item {\texttt{soprano.analyse}}, \pageref{soprano.analyse:module-soprano.analyse}
\item {\texttt{soprano.analyse.phylogen}}, \pageref{soprano.analyse.phylogen:module-soprano.analyse.phylogen}
\item {\texttt{soprano.analyse.phylogen.genes}}, \pageref{soprano.analyse.phylogen.genes:module-soprano.analyse.phylogen.genes}
\item {\texttt{soprano.analyse.phylogen.mapping}}, \pageref{soprano.analyse.phylogen.mapping:module-soprano.analyse.phylogen.mapping}
\item {\texttt{soprano.analyse.phylogen.phylogenclust}}, \pageref{soprano.analyse.phylogen.phylogenclust:module-soprano.analyse.phylogen.phylogenclust}
\item {\texttt{soprano.calculate}}, \pageref{soprano.calculate:module-soprano.calculate}
\item {\texttt{soprano.calculate.gulp}}, \pageref{soprano.calculate.gulp:module-soprano.calculate.gulp}
\item {\texttt{soprano.calculate.gulp.charges}}, \pageref{soprano.calculate.gulp.charges:module-soprano.calculate.gulp.charges}
\item {\texttt{soprano.calculate.gulp.w99}}, \pageref{soprano.calculate.gulp.w99:module-soprano.calculate.gulp.w99}
\item {\texttt{soprano.calculate.xrd}}, \pageref{soprano.calculate.xrd:module-soprano.calculate.xrd}
\item {\texttt{soprano.calculate.xrd.sel\_rules}}, \pageref{soprano.calculate.xrd.sel_rules:module-soprano.calculate.xrd.sel_rules}
\item {\texttt{soprano.calculate.xrd.xrd}}, \pageref{soprano.calculate.xrd.xrd:module-soprano.calculate.xrd.xrd}
\item {\texttt{soprano.collection}}, \pageref{soprano.collection:module-soprano.collection}
\item {\texttt{soprano.collection.collection}}, \pageref{soprano.collection.collection:module-soprano.collection.collection}
\item {\texttt{soprano.collection.generate}}, \pageref{soprano.collection.generate:module-soprano.collection.generate}
\item {\texttt{soprano.collection.generate.airss}}, \pageref{soprano.collection.generate.airss:module-soprano.collection.generate.airss}
\item {\texttt{soprano.collection.generate.linspace}}, \pageref{soprano.collection.generate.linspace:module-soprano.collection.generate.linspace}
\item {\texttt{soprano.collection.generate.rattle}}, \pageref{soprano.collection.generate.rattle:module-soprano.collection.generate.rattle}
\item {\texttt{soprano.hpc}}, \pageref{soprano.hpc:module-soprano.hpc}
\item {\texttt{soprano.hpc.submitter}}, \pageref{soprano.hpc.submitter:module-soprano.hpc.submitter}
\item {\texttt{soprano.hpc.submitter.castep}}, \pageref{soprano.hpc.submitter.castep:module-soprano.hpc.submitter.castep}
\item {\texttt{soprano.hpc.submitter.debug}}, \pageref{soprano.hpc.submitter.debug:module-soprano.hpc.submitter.debug}
\item {\texttt{soprano.hpc.submitter.debug.debugqueue}}, \pageref{soprano.hpc.submitter.debug.debugqueue:module-soprano.hpc.submitter.debug.debugqueue}
\item {\texttt{soprano.hpc.submitter.queues}}, \pageref{soprano.hpc.submitter.queues:module-soprano.hpc.submitter.queues}
\item {\texttt{soprano.hpc.submitter.submit}}, \pageref{soprano.hpc.submitter.submit:module-soprano.hpc.submitter.submit}
\item {\texttt{soprano.properties}}, \pageref{soprano.properties:module-soprano.properties}
\item {\texttt{soprano.properties.atomsproperty}}, \pageref{soprano.properties.atomsproperty:module-soprano.properties.atomsproperty}
\item {\texttt{soprano.properties.basic}}, \pageref{soprano.properties.basic:module-soprano.properties.basic}
\item {\texttt{soprano.properties.basic.basic}}, \pageref{soprano.properties.basic.basic:module-soprano.properties.basic.basic}
\item {\texttt{soprano.properties.castep}}, \pageref{soprano.properties.castep:module-soprano.properties.castep}
\item {\texttt{soprano.properties.castep.castep}}, \pageref{soprano.properties.castep.castep:module-soprano.properties.castep.castep}
\item {\texttt{soprano.properties.labeling}}, \pageref{soprano.properties.labeling:module-soprano.properties.labeling}
\item {\texttt{soprano.properties.labeling.labeling}}, \pageref{soprano.properties.labeling.labeling:module-soprano.properties.labeling.labeling}
\item {\texttt{soprano.properties.linkage}}, \pageref{soprano.properties.linkage:module-soprano.properties.linkage}
\item {\texttt{soprano.properties.linkage.linkage}}, \pageref{soprano.properties.linkage.linkage:module-soprano.properties.linkage.linkage}
\item {\texttt{soprano.properties.symmetry}}, \pageref{soprano.properties.symmetry:module-soprano.properties.symmetry}
\item {\texttt{soprano.properties.symmetry.symmetry}}, \pageref{soprano.properties.symmetry.symmetry:module-soprano.properties.symmetry.symmetry}
\item {\texttt{soprano.properties.transform}}, \pageref{soprano.properties.transform:module-soprano.properties.transform}
\item {\texttt{soprano.properties.transform.transform}}, \pageref{soprano.properties.transform.transform:module-soprano.properties.transform.transform}
\item {\texttt{soprano.selection}}, \pageref{soprano.selection:module-soprano.selection}
\item {\texttt{soprano.utils}}, \pageref{soprano.utils:module-soprano.utils}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
