
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>soprano.analyse.phylogen.phylogenclust &#8212; Soprano</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script >const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
    <script >mermaid.initialize({startOnLoad:true});</script>
    <link rel="shortcut icon" href="../_static/favicon.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="soprano.calculate" href="soprano.calculate.html" />
    <link rel="prev" title="soprano.analyse.phylogen.metrics" href="soprano.analyse.phylogen.metrics.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/soprano_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Soprano</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    About
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../tutorials.html">
   Tutorials
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/01-basic_concepts.html">
     Tutorial 1 - Basic concepts: using AtomsCollection objects
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/02-generators_properties_calculators.html">
     Tutorial 2 - Generators, Properties and Calculators
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/03-atomselection_transforms.html">
     Tutorial 3 - AtomSelection and transforms
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/04-clustering.html">
     Tutorial 4 - Clustering analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/05-nmr.html">
     Tutorial 5 - NMR Properties: using Soprano for working with .magres files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../tutorials/06-defect_calculations.html">
     Tutorial 6 - Soprano for defect calculations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../submitter.html">
   Job Submitter
  </a>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../api.html">
   API reference
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active has-children">
    <a class="reference internal" href="soprano.html">
     soprano
    </a>
    <input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul class="current">
     <li class="toctree-l3 current active has-children">
      <a class="reference internal" href="soprano.analyse.html">
       soprano.analyse
      </a>
      <input checked="" class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
      <label for="toctree-checkbox-4">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul class="current">
       <li class="toctree-l4 current active">
        <a class="reference internal" href="soprano.analyse.phylogen.html">
         soprano.analyse.phylogen
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.calculate.html">
       soprano.calculate
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
      <label for="toctree-checkbox-5">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.calculate.gulp.html">
         soprano.calculate.gulp
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.calculate.nmr.html">
         soprano.calculate.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.calculate.powder.html">
         soprano.calculate.powder
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.calculate.xrd.html">
         soprano.calculate.xrd
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.collection.html">
       soprano.collection
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
      <label for="toctree-checkbox-6">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.collection.collection.html">
         soprano.collection.collection
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.collection.generate.html">
         soprano.collection.generate
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.data.html">
       soprano.data
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
      <label for="toctree-checkbox-7">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.data.nmr.html">
         soprano.data.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.data.vdw.html">
         soprano.data.vdw
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.hpc.html">
       soprano.hpc
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
      <label for="toctree-checkbox-8">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.hpc.submitter.html">
         soprano.hpc.submitter
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="soprano.measure.html">
       soprano.measure
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.nmr.html">
       soprano.nmr
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
      <label for="toctree-checkbox-9">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.nmr.tensor.html">
         soprano.nmr.tensor
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.nmr.utils.html">
         soprano.nmr.utils
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="soprano.optional.html">
       soprano.optional
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.properties.html">
       soprano.properties
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
      <label for="toctree-checkbox-10">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.atomsproperty.html">
         soprano.properties.atomsproperty
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.basic.html">
         soprano.properties.basic
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.castep.html">
         soprano.properties.castep
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.labeling.html">
         soprano.properties.labeling
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.linkage.html">
         soprano.properties.linkage
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.map.html">
         soprano.properties.map
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.nmr.html">
         soprano.properties.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.order.html">
         soprano.properties.order
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.symmetry.html">
         soprano.properties.symmetry
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.properties.transform.html">
         soprano.properties.transform
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="soprano.rnd.html">
       soprano.rnd
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="soprano.scripts.html">
       soprano.scripts
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
      <label for="toctree-checkbox-11">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.scripts.msaverage.html">
         soprano.scripts.msaverage
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.scripts.phylogen.html">
         soprano.scripts.phylogen
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="soprano.scripts.vasp2cell.html">
         soprano.scripts.vasp2cell
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="soprano.selection.html">
       soprano.selection
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="soprano.utils.html">
       soprano.utils
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../citing.html">
   Citing Soprano
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            A Python library to crack crystals
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/CCP-NC/soprano"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/CCP-NC/soprano/issues/new?title=Issue%20on%20page%20%2F_autosummary/soprano.analyse.phylogen.phylogenclust.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/_autosummary/soprano.analyse.phylogen.phylogenclust.rst"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.rst</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>soprano.analyse.phylogen.phylogenclust</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="module-soprano.analyse.phylogen.phylogenclust">
<span id="soprano-analyse-phylogen-phylogenclust"></span><h1>soprano.analyse.phylogen.phylogenclust<a class="headerlink" href="#module-soprano.analyse.phylogen.phylogenclust" title="Permalink to this headline">¶</a></h1>
<p>Phylogenetic clustering class definitions</p>
<p class="rubric">Classes</p>
<table class="longtable table autosummary">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster" title="soprano.analyse.phylogen.phylogenclust.PhylogenCluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PhylogenCluster</span></code></a>(coll[, genes, norm_range, …])</p></td>
<td><p>An object that, given an AtomsCollection and a series of “genes” and weights, will build clusters out of the structures in the collection based on their reciprocal positions as points in a multi-dimensional space defined by those “genes”.</p></td>
</tr>
</tbody>
</table>
<dl class="py class">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster">
<em class="property"><span class="pre">class</span> </em><code class="sig-prename descclassname"><span class="pre">soprano.analyse.phylogen.phylogenclust.</span></code><code class="sig-name descname"><span class="pre">PhylogenCluster</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coll</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.0,</span> <span class="pre">1.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An object that, given an AtomsCollection and a series of “genes” and
weights, will build clusters out of the structures in the collection based
on their reciprocal positions as points in a multi-dimensional space
defined by those “genes”.</p>
<p>Initialize the PhylogenCluster object.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">coll (AtomsCollection): an AtomsCollection containing the</div>
<div class="line-block">
<div class="line">structures that should be classified.</div>
<div class="line">This will be copied and frozen for the</div>
<div class="line">entirety of the life of this instance;</div>
<div class="line">in order to operate on a modified</div>
<div class="line">collection, a new PhylogenCluster should</div>
<div class="line">be created.</div>
</div>
<div class="line">genes (list[tuple], str, file): list of the genes that should be</div>
<div class="line-block">
<div class="line">loaded immediately; each gene</div>
<div class="line">comes in the form of a tuple</div>
<div class="line">(name (str), weight (float),</div>
<div class="line">params (dict)). A path or open</div>
<div class="line">file can also be passed for a</div>
<div class="line">.gene file, from which the values</div>
<div class="line">will be loaded.</div>
</div>
<div class="line">norm_range (list[float?]): ranges to constrain the values of</div>
<div class="line-block">
<div class="line">single genes in between. Default is</div>
<div class="line">(0, 1). A value of “None” in either</div>
<div class="line">place can be used to indicate no</div>
<div class="line">normalization on one or both sides.</div>
</div>
<div class="line">norm_dist (float?): value to normalize distance genes to. These</div>
<div class="line-block">
<div class="line">are the genes that only make sense on pairs of</div>
<div class="line">structures. Their minimum value is always 0.</div>
<div class="line">This number would become their maximum value,</div>
<div class="line">or can be set to None to avoid normalization.</div>
</div>
</div>
</div>
<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc">
<code class="sig-name descname"><span class="pre">_recalc</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster._recalc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc" title="Permalink to this definition">¶</a></dt>
<dd><p>Recalculate all stored variables that depend on genes and ranges</p>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping">
<code class="sig-name descname"><span class="pre">create_mapping</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'total-principal'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.create_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an array of 2-dimensional points representing a reduced
dimensionality mapping of the given genes using the algorithm of
choice. All algorithms are described in [W. Siedlecki et al., Patt.
Recog. vol. 21, num. 5, pp. 411 429 (1988)].</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): can be one of the following algorithms:</div>
<div class="line-block">
<div class="line">- total_principal (default)</div>
<div class="line">- clafic</div>
<div class="line">- fukunaga-koontz</div>
<div class="line">- optimal-discriminant</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats">
<code class="sig-name descname"><span class="pre">get_cluster_stats</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_cluster_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute average values and standard deviation for each gene within
a given clustering.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">clusters (tuple): the clustering in tuple form, as returned by one</div>
<div class="line-block">
<div class="line">of the get_clusters methods.</div>
</div>
<div class="line">raw (bool): if True, return average and standard deviation of raw</div>
<div class="line-block">
<div class="line">instead of normalised gene values. Default is False.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">avgs (np.ndarray): 2D array of average values of each gene for</div>
<div class="line-block">
<div class="line">each cluster.</div>
</div>
<div class="line">stds (np.ndarray): 2D array of standard deviations of each gene</div>
<div class="line-block">
<div class="line">for each cluster.</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">arrays</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters">
<code class="sig-name descname"><span class="pre">get_clusters</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Wrapper method to get clusters by any available method. Depending
on the value passed as ‘method’ it calls either ger_hier_clusters,
get_kmeans_clusters, or get_sklearn_clusters. Check their respective
docstrings for more detailed info.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering method to use. Can be ‘hier’,</div>
<div class="line-block">
<div class="line">‘kmeans’, or one of the methods in sklearn.clusters.</div>
</div>
<div class="line">params (dict): parameters to be passed to the class when</div>
<div class="line-block">
<div class="line">initialising it. Change depending on the desired</div>
<div class="line">method. Check the documentation for the specific</div>
<div class="line">class.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by the</div>
<div class="line">requested algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat">
<code class="sig-name descname"><span class="pre">get_distmat</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_distmat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the distance matrix between structures in the collection,
based on the genes currently in use.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">distmat (np.ndarray): a (collection.length, collection.length)</div>
<div class="line-block">
<div class="line">array, containing the overall distance</div>
<div class="line">(the norm of all individual gene distances)</div>
<div class="line">between all pairs of structures.</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot">
<code class="sig-name descname"><span class="pre">get_elbow_plot</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kmeans'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'n'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">range(1,</span> <span class="pre">11)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_elbow_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns data for an elbow plot by scanning the outcome of a given
clustering method within a range of values for a chosen parameter.
Used to determine optimal parameter values.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering method to use. Can be ‘hier’,</div>
<div class="line-block">
<div class="line">‘kmeans’, or one of the methods in sklearn.clusters.</div>
<div class="line">Default is kmeans.</div>
</div>
<div class="line">param_name (str): parameter to be scanned over. Change depending</div>
<div class="line-block">
<div class="line">on the desired method. Check the documentation</div>
<div class="line">for the specific class. Default is n, number of</div>
<div class="line">clusters for k-means method.</div>
</div>
<div class="line">param_range (list): values of param_name to scan over. Default is</div>
<div class="line-block">
<div class="line">the integers from 1 to 10.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">wss (np.ndarray): values of the “Within cluster Sum of Squares”</div>
<div class="line-block">
<div class="line">(WSS) to be used on the elbow plot y axis.</div>
</div>
<div class="line">param_range (list): range used for parameter scan, to be used on</div>
<div class="line-block">
<div class="line">the x axis (same as passed by the user).</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices">
<code class="sig-name descname"><span class="pre">get_genome_matrices</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_matrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genome matrices in raw form (not normalized).
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_matrix (np.ndarray): a (collection.length,</div>
<div class="line-block">
<div class="line">collection.length, gene.length)</div>
<div class="line">array, containing the distances for</div>
<div class="line">each gene and pair of structures in</div>
<div class="line">row and column</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm">
<code class="sig-name descname"><span class="pre">get_genome_matrices_norm</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_matrices_norm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genome matrices in normalized and weighted form.
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_matrix (np.ndarray): a (collection.length,</div>
<div class="line-block">
<div class="line">collection.length, gene.length)</div>
<div class="line">array, containing the distances for</div>
<div class="line">each gene and pair of structures in</div>
<div class="line">row and column</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors">
<code class="sig-name descname"><span class="pre">get_genome_vectors</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_vectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genome vectors in raw form (not normalized).
The vectors refer to genes that allow to define a specific point for
each structure.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_vectors (np.ndarray): a (collection.length, gene.length)</div>
<div class="line-block">
<div class="line">array, containing the whole extent</div>
<div class="line">of the gene values for each structure</div>
<div class="line">in the collection on each row</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm">
<code class="sig-name descname"><span class="pre">get_genome_vectors_norm</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_vectors_norm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the genome vectors in normalized and weighted form.
The vectors refer to genes that allow to define a specific point for
each structure.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_vectors (np.ndarray): a (collection.length, gene.length)</div>
<div class="line-block">
<div class="line">array, containing the whole extent</div>
<div class="line">of the gene values for each structure</div>
<div class="line">in the collection on each row</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters">
<code class="sig-name descname"><span class="pre">get_hier_clusters</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_hier_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Get multiple clusters (in the form of a list of collections) based
on the hierarchical clustering methods and the currently set genes.</p>
<p>Calls scipy.cluster.hierarchy.fcluster</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">t (float): minimum distance of separation required to consider</div>
<div class="line-block">
<div class="line">two clusters separate. This controls the number of</div>
<div class="line">clusters: a smaller value will produce more fine</div>
<div class="line">grained clustering. At the limit, a value smaller than</div>
<div class="line">the distance between the two closest structures will</div>
<div class="line">return a cluster for each structure. Remember that the</div>
<div class="line">‘distances’ in this case refer to distances between the</div>
<div class="line">‘gene’ values attributed to each structure. In other</div>
<div class="line">words they are a function of the chosen genes,</div>
<div class="line">normalization conditions and weights employed.</div>
<div class="line">In addition, the way they are calculated depends on the</div>
<div class="line">choice of method.</div>
</div>
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by hierarchical</div>
<div class="line">algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree">
<code class="sig-name descname"><span class="pre">get_hier_tree</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_hier_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a tree data structure describing the clustering order of based
on the hierarchical clustering methods and the currently set genes.</p>
<p>Calls scipy.cluster.hierarchy.to_tree</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">root_node (ClusterNode): the root node of the tree. Access child</div>
<div class="line-block">
<div class="line">members with .left and .right, while .id</div>
<div class="line">holds the number of the corresponding</div>
<div class="line">cluster. Refer to Scipy documentation for</div>
<div class="line">further details.</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters">
<code class="sig-name descname"><span class="pre">get_kmeans_clusters</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_kmeans_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a given number of clusters (in the form of a list of
collections) based on the k-means clustering methods
and the currently set genes.
Warning: this method only works if there are no genes that work only
with pairs of structures - as specific points, and not just distances
between them, are required for this algorithm.</p>
<p>Calls scipy.cluster.vq.kmeans</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">n (int):    the desired number of clusters.</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by k-means</div>
<div class="line">algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage">
<code class="sig-name descname"><span class="pre">get_linkage</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_linkage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the linkage matrix between structures in the collection,
based on the genes currently in use. Only used in hierarchical
clustering.</p>
<p>Calls scipy.cluster.hierarchy.linkage.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">Z (np.ndarray): linkage matrix for the structures in the</div>
<div class="line-block">
<div class="line">collection. Refer to Scipy documentation for</div>
<div class="line">details about the method</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist">
<code class="sig-name descname"><span class="pre">get_max_cluster_dist</span></code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_max_cluster_dist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the maximum possible distance between two clusters</p>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters">
<code class="sig-name descname"><span class="pre">get_sklearn_clusters</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_sklearn_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>Get clusters applying any of the methods provided by the library
scikit-learn (requires a separate installation).
Warning: this method only works if there are no genes that work only
with pairs of structures - as use of pairwise clustering methods is
not implemented yet.</p>
<p>Uses the sklearn.cluster.&lt;method&gt; class</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering class from sklearn.clusters</div>
<div class="line-block">
<div class="line">to use. For reference check the documentation at</div>
<div class="line"><a class="reference external" href="http://scikit-learn.org/stable/modules/clustering.html">http://scikit-learn.org/stable/modules/clustering.html</a></div>
</div>
<div class="line">params (dict): parameters to be passed to the class when</div>
<div class="line-block">
<div class="line">initialising it. Change depending on the desired</div>
<div class="line">method. Check the documentation for the specific</div>
<div class="line">class.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by the</div>
<div class="line">requested algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load">
<em class="property"><span class="pre">static</span> </em><code class="sig-name descname"><span class="pre">load</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a pickled copy from a given file path</p>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save">
<code class="sig-name descname"><span class="pre">save</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Simply save a pickled copy to a given file path</p>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection">
<code class="sig-name descname"><span class="pre">save_collection</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.save_collection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection" title="Permalink to this definition">¶</a></dt>
<dd><p>Save as pickle the collection bound to this PhylogenCluster.
The calculated genes are also stored in it as arrays for future use.</p>
</dd></dl>

<dl class="py method">
<dt id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes">
<code class="sig-name descname"><span class="pre">set_genes</span></code><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_arrays</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.set_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate, store and set a list of genes as used for clustering.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">genes (list[soprano.analyse.phylogen.Gene],</div>
<div class="line-block">
<div class="line">file, str): a list of Genes to calculate and store. A path</div>
<div class="line-block">
<div class="line">or open file can also be passed for a .gene</div>
<div class="line">file, from which the values will be loaded.</div>
</div>
</div>
<div class="line">load_arrays (bool): try loading the genes as arrays from the</div>
<div class="line-block">
<div class="line">collection before generating them. Warning:</div>
<div class="line">if there are arrays named like genes but with</div>
<div class="line">different contents this can lead to</div>
<div class="line">unpredictable results.</div>
</div>
</div>
</div>
</dd></dl>

</dd></dl>

</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./_autosummary"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="soprano.analyse.phylogen.metrics.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">soprano.analyse.phylogen.metrics</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="soprano.calculate.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">soprano.calculate</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By CCP-NC<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>