
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>soprano.analyse.phylogen.phylogenclust &#8212; Soprano</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-design.min.css?v=87e54e7c" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=5eb1bb1e"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
    <script>mermaid.initialize({startOnLoad:true});</script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_autosummary/soprano.analyse.phylogen.phylogenclust';</script>
    <link rel="icon" href="../_static/favicon.png"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="soprano.calculate" href="soprano.calculate.html" />
    <link rel="prev" title="soprano.analyse.phylogen.metrics" href="soprano.analyse.phylogen.metrics.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../_static/soprano_logo.svg" class="logo__image only-light" alt="Soprano - Home"/>
    <script>document.write(`<img src="../_static/soprano_logo.svg" class="logo__image only-dark" alt="Soprano - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../intro.html">
                    About
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/01-basic_concepts.html">Tutorial 1 - Basic concepts: using AtomsCollection objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/02-generators_properties_calculators.html">Tutorial 2 - Generators, Properties and Calculators</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/03-atomselection_transforms.html">Tutorial 3 - AtomSelection and transforms</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/04-clustering.html">Tutorial 4 - Clustering analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/05-nmr.html">Tutorial 5 - NMR Properties: using Soprano for working with .magres files</a></li>




<li class="toctree-l2"><a class="reference internal" href="../tutorials/06-defect_calculations.html">Tutorial 6 - Soprano for defect calculations</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../submitter.html">Job Submitter</a></li>

<li class="toctree-l1 current active has-children"><a class="reference internal" href="../api.html">API reference</a><ul class="current">
<li class="toctree-l2 current active has-children"><a class="reference internal" href="soprano.html">soprano</a><ul class="current">
<li class="toctree-l3 current active has-children"><a class="reference internal" href="soprano.analyse.html">soprano.analyse</a><ul class="current">
<li class="toctree-l4 current active"><a class="reference internal" href="soprano.analyse.phylogen.html">soprano.analyse.phylogen</a></li>
</ul><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.calculate.html">soprano.calculate</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.calculate.gulp.html">soprano.calculate.gulp</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.calculate.nmr.html">soprano.calculate.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.calculate.powder.html">soprano.calculate.powder</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.calculate.xrd.html">soprano.calculate.xrd</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.collection.html">soprano.collection</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.collection.collection.html">soprano.collection.collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.collection.generate.html">soprano.collection.generate</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.data.html">soprano.data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.data.nmr.html">soprano.data.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.data.vdw.html">soprano.data.vdw</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.hpc.html">soprano.hpc</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.hpc.submitter.html">soprano.hpc.submitter</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="soprano.measure.html">soprano.measure</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.nmr.html">soprano.nmr</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.nmr.tensor.html">soprano.nmr.tensor</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.nmr.utils.html">soprano.nmr.utils</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="soprano.optional.html">soprano.optional</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.properties.html">soprano.properties</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.atomsproperty.html">soprano.properties.atomsproperty</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.basic.html">soprano.properties.basic</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.castep.html">soprano.properties.castep</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.labeling.html">soprano.properties.labeling</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.linkage.html">soprano.properties.linkage</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.map.html">soprano.properties.map</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.nmr.html">soprano.properties.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.order.html">soprano.properties.order</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.symmetry.html">soprano.properties.symmetry</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.properties.transform.html">soprano.properties.transform</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="soprano.rnd.html">soprano.rnd</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="soprano.scripts.html">soprano.scripts</a><ul>
<li class="toctree-l4"><a class="reference internal" href="soprano.scripts.msaverage.html">soprano.scripts.msaverage</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.scripts.phylogen.html">soprano.scripts.phylogen</a></li>
<li class="toctree-l4"><a class="reference internal" href="soprano.scripts.vasp2cell.html">soprano.scripts.vasp2cell</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="soprano.selection.html">soprano.selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="soprano.utils.html">soprano.utils</a></li>
</ul><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
</ul><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>

<li class="toctree-l1"><a class="reference internal" href="../citing.html">Citing Soprano</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/CCP-NC/soprano" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/CCP-NC/soprano/issues/new?title=Issue%20on%20page%20%2F_autosummary/soprano.analyse.phylogen.phylogenclust.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/_autosummary/soprano.analyse.phylogen.phylogenclust.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>soprano.analyse.phylogen.phylogenclust</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster._recalc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.create_mapping()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_cluster_stats()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_distmat()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_elbow_plot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_matrices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_matrices_norm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_vectors()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_vectors_norm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_hier_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_hier_tree()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_kmeans_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_linkage()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_max_cluster_dist()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_sklearn_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.load()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.save()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.save_collection()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.set_genes()</span></code></a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="module-soprano.analyse.phylogen.phylogenclust">
<span id="soprano-analyse-phylogen-phylogenclust"></span><h1>soprano.analyse.phylogen.phylogenclust<a class="headerlink" href="#module-soprano.analyse.phylogen.phylogenclust" title="Link to this heading">#</a></h1>
<p>Phylogenetic clustering class definitions</p>
<p class="rubric">Classes</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster" title="soprano.analyse.phylogen.phylogenclust.PhylogenCluster"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PhylogenCluster</span></code></a>(coll[, genes, norm_range, ...])</p></td>
<td><p>An object that, given an AtomsCollection and a series of &quot;genes&quot; and weights, will build clusters out of the structures in the collection based on their reciprocal positions as points in a multi-dimensional space defined by those &quot;genes&quot;.</p></td>
</tr>
</tbody>
</table>
</div>
<dl class="py class">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">soprano.analyse.phylogen.phylogenclust.</span></span><span class="sig-name descname"><span class="pre">PhylogenCluster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coll</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0.0,</span> <span class="pre">1.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">norm_dist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster" title="Link to this definition">#</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An object that, given an AtomsCollection and a series of “genes” and
weights, will build clusters out of the structures in the collection based
on their reciprocal positions as points in a multi-dimensional space
defined by those “genes”.</p>
<p>Initialize the PhylogenCluster object.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">coll (AtomsCollection): an AtomsCollection containing the</div>
<div class="line-block">
<div class="line">structures that should be classified.</div>
<div class="line">This will be copied and frozen for the</div>
<div class="line">entirety of the life of this instance;</div>
<div class="line">in order to operate on a modified</div>
<div class="line">collection, a new PhylogenCluster should</div>
<div class="line">be created.</div>
</div>
<div class="line">genes (list[tuple], str, file): list of the genes that should be</div>
<div class="line-block">
<div class="line">loaded immediately; each gene</div>
<div class="line">comes in the form of a tuple</div>
<div class="line">(name (str), weight (float),</div>
<div class="line">params (dict)). A path or open</div>
<div class="line">file can also be passed for a</div>
<div class="line">.gene file, from which the values</div>
<div class="line">will be loaded.</div>
</div>
<div class="line">norm_range (list[float?]): ranges to constrain the values of</div>
<div class="line-block">
<div class="line">single genes in between. Default is</div>
<div class="line">(0, 1). A value of “None” in either</div>
<div class="line">place can be used to indicate no</div>
<div class="line">normalization on one or both sides.</div>
</div>
<div class="line">norm_dist (float?): value to normalize distance genes to. These</div>
<div class="line-block">
<div class="line">are the genes that only make sense on pairs of</div>
<div class="line">structures. Their minimum value is always 0.</div>
<div class="line">This number would become their maximum value,</div>
<div class="line">or can be set to None to avoid normalization.</div>
</div>
</div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc">
<span class="sig-name descname"><span class="pre">_recalc</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster._recalc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc" title="Link to this definition">#</a></dt>
<dd><p>Recalculate all stored variables that depend on genes and ranges</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping">
<span class="sig-name descname"><span class="pre">create_mapping</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'total-principal'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.create_mapping"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping" title="Link to this definition">#</a></dt>
<dd><p>Return an array of 2-dimensional points representing a reduced
dimensionality mapping of the given genes using the algorithm of
choice. All algorithms are described in [W. Siedlecki et al., Patt.
Recog. vol. 21, num. 5, pp. 411 429 (1988)].</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): can be one of the following algorithms:</div>
<div class="line-block">
<div class="line">- total_principal (default)</div>
<div class="line">- clafic</div>
<div class="line">- fukunaga-koontz</div>
<div class="line">- optimal-discriminant</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats">
<span class="sig-name descname"><span class="pre">get_cluster_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">clusters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raw</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_cluster_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats" title="Link to this definition">#</a></dt>
<dd><p>Compute average values and standard deviation for each gene within
a given clustering.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">clusters (tuple): the clustering in tuple form, as returned by one</div>
<div class="line-block">
<div class="line">of the get_clusters methods.</div>
</div>
<div class="line">raw (bool): if True, return average and standard deviation of raw</div>
<div class="line-block">
<div class="line">instead of normalised gene values. Default is False.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">avgs (np.ndarray): 2D array of average values of each gene for</div>
<div class="line-block">
<div class="line">each cluster.</div>
</div>
<div class="line">stds (np.ndarray): 2D array of standard deviations of each gene</div>
<div class="line-block">
<div class="line">for each cluster.</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">arrays</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters">
<span class="sig-name descname"><span class="pre">get_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters" title="Link to this definition">#</a></dt>
<dd><p>Wrapper method to get clusters by any available method. Depending
on the value passed as ‘method’ it calls either ger_hier_clusters,
get_kmeans_clusters, or get_sklearn_clusters. Check their respective
docstrings for more detailed info.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering method to use. Can be ‘hier’,</div>
<div class="line-block">
<div class="line">‘kmeans’, or one of the methods in sklearn.clusters.</div>
</div>
<div class="line">params (dict): parameters to be passed to the class when</div>
<div class="line-block">
<div class="line">initialising it. Change depending on the desired</div>
<div class="line">method. Check the documentation for the specific</div>
<div class="line">class.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by the</div>
<div class="line">requested algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat">
<span class="sig-name descname"><span class="pre">get_distmat</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_distmat"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat" title="Link to this definition">#</a></dt>
<dd><p>Get the distance matrix between structures in the collection,
based on the genes currently in use.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">distmat (np.ndarray): a (collection.length, collection.length)</div>
<div class="line-block">
<div class="line">array, containing the overall distance</div>
<div class="line">(the norm of all individual gene distances)</div>
<div class="line">between all pairs of structures.</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot">
<span class="sig-name descname"><span class="pre">get_elbow_plot</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'kmeans'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'n'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">param_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">range(1,</span> <span class="pre">11)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_elbow_plot"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot" title="Link to this definition">#</a></dt>
<dd><p>Returns data for an elbow plot by scanning the outcome of a given
clustering method within a range of values for a chosen parameter.
Used to determine optimal parameter values.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering method to use. Can be ‘hier’,</div>
<div class="line-block">
<div class="line">‘kmeans’, or one of the methods in sklearn.clusters.</div>
<div class="line">Default is kmeans.</div>
</div>
<div class="line">param_name (str): parameter to be scanned over. Change depending</div>
<div class="line-block">
<div class="line">on the desired method. Check the documentation</div>
<div class="line">for the specific class. Default is n, number of</div>
<div class="line">clusters for k-means method.</div>
</div>
<div class="line">param_range (list): values of param_name to scan over. Default is</div>
<div class="line-block">
<div class="line">the integers from 1 to 10.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">wss (np.ndarray): values of the “Within cluster Sum of Squares”</div>
<div class="line-block">
<div class="line">(WSS) to be used on the elbow plot y axis.</div>
</div>
<div class="line">param_range (list): range used for parameter scan, to be used on</div>
<div class="line-block">
<div class="line">the x axis (same as passed by the user).</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices">
<span class="sig-name descname"><span class="pre">get_genome_matrices</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_matrices"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices" title="Link to this definition">#</a></dt>
<dd><p>Return the genome matrices in raw form (not normalized).
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_matrix (np.ndarray): a (collection.length,</div>
<div class="line-block">
<div class="line">collection.length, gene.length)</div>
<div class="line">array, containing the distances for</div>
<div class="line">each gene and pair of structures in</div>
<div class="line">row and column</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm">
<span class="sig-name descname"><span class="pre">get_genome_matrices_norm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_matrices_norm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm" title="Link to this definition">#</a></dt>
<dd><p>Return the genome matrices in normalized and weighted form.
The matrices refer to genes that only allow to define a distance
between structures. The element at i,j represents the distance
between said structures. The matrix is symmetric and has
null diagonal.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_matrix (np.ndarray): a (collection.length,</div>
<div class="line-block">
<div class="line">collection.length, gene.length)</div>
<div class="line">array, containing the distances for</div>
<div class="line">each gene and pair of structures in</div>
<div class="line">row and column</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors">
<span class="sig-name descname"><span class="pre">get_genome_vectors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_vectors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors" title="Link to this definition">#</a></dt>
<dd><p>Return the genome vectors in raw form (not normalized).
The vectors refer to genes that allow to define a specific point for
each structure.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_vectors (np.ndarray): a (collection.length, gene.length)</div>
<div class="line-block">
<div class="line">array, containing the whole extent</div>
<div class="line">of the gene values for each structure</div>
<div class="line">in the collection on each row</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm">
<span class="sig-name descname"><span class="pre">get_genome_vectors_norm</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_genome_vectors_norm"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm" title="Link to this definition">#</a></dt>
<dd><p>Return the genome vectors in normalized and weighted form.
The vectors refer to genes that allow to define a specific point for
each structure.</p>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">genome_vectors (np.ndarray): a (collection.length, gene.length)</div>
<div class="line-block">
<div class="line">array, containing the whole extent</div>
<div class="line">of the gene values for each structure</div>
<div class="line">in the collection on each row</div>
</div>
<div class="line">genome_legend (list[tuple]): a list of tuples containing (name,</div>
<div class="line-block">
<div class="line">length) of the gene fragments in the</div>
<div class="line">array</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters">
<span class="sig-name descname"><span class="pre">get_hier_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_hier_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters" title="Link to this definition">#</a></dt>
<dd><p>Get multiple clusters (in the form of a list of collections) based
on the hierarchical clustering methods and the currently set genes.</p>
<p>Calls scipy.cluster.hierarchy.fcluster</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">t (float): minimum distance of separation required to consider</div>
<div class="line-block">
<div class="line">two clusters separate. This controls the number of</div>
<div class="line">clusters: a smaller value will produce more fine</div>
<div class="line">grained clustering. At the limit, a value smaller than</div>
<div class="line">the distance between the two closest structures will</div>
<div class="line">return a cluster for each structure. Remember that the</div>
<div class="line">‘distances’ in this case refer to distances between the</div>
<div class="line">‘gene’ values attributed to each structure. In other</div>
<div class="line">words they are a function of the chosen genes,</div>
<div class="line">normalization conditions and weights employed.</div>
<div class="line">In addition, the way they are calculated depends on the</div>
<div class="line">choice of method.</div>
</div>
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by hierarchical</div>
<div class="line">algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree">
<span class="sig-name descname"><span class="pre">get_hier_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_hier_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree" title="Link to this definition">#</a></dt>
<dd><p>Get a tree data structure describing the clustering order of based
on the hierarchical clustering methods and the currently set genes.</p>
<p>Calls scipy.cluster.hierarchy.to_tree</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">root_node (ClusterNode): the root node of the tree. Access child</div>
<div class="line-block">
<div class="line">members with .left and .right, while .id</div>
<div class="line">holds the number of the corresponding</div>
<div class="line">cluster. Refer to Scipy documentation for</div>
<div class="line">further details.</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters">
<span class="sig-name descname"><span class="pre">get_kmeans_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_kmeans_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters" title="Link to this definition">#</a></dt>
<dd><p>Get a given number of clusters (in the form of a list of
collections) based on the k-means clustering methods
and the currently set genes.
Warning: this method only works if there are no genes that work only
with pairs of structures - as specific points, and not just distances
between them, are required for this algorithm.</p>
<p>Calls scipy.cluster.vq.kmeans</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">n (int):    the desired number of clusters.</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by k-means</div>
<div class="line">algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage">
<span class="sig-name descname"><span class="pre">get_linkage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'single'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_linkage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage" title="Link to this definition">#</a></dt>
<dd><p>Get the linkage matrix between structures in the collection,
based on the genes currently in use. Only used in hierarchical
clustering.</p>
<p>Calls scipy.cluster.hierarchy.linkage.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): clustering method to employ. Valid entries are</div>
<div class="line-block">
<div class="line">‘single’, ‘complete’, ‘weighted’ and ‘average’.</div>
<div class="line">Refer to Scipy documentation for further details.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">Z (np.ndarray): linkage matrix for the structures in the</div>
<div class="line-block">
<div class="line">collection. Refer to Scipy documentation for</div>
<div class="line">details about the method</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist">
<span class="sig-name descname"><span class="pre">get_max_cluster_dist</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_max_cluster_dist"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist" title="Link to this definition">#</a></dt>
<dd><p>Return the maximum possible distance between two clusters</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters">
<span class="sig-name descname"><span class="pre">get_sklearn_clusters</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.get_sklearn_clusters"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters" title="Link to this definition">#</a></dt>
<dd><p>Get clusters applying any of the methods provided by the library
scikit-learn (requires a separate installation).
Warning: this method only works if there are no genes that work only
with pairs of structures - as use of pairwise clustering methods is
not implemented yet.</p>
<p>Uses the sklearn.cluster.&lt;method&gt; class</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">method (str): name of the clustering class from sklearn.clusters</div>
<div class="line-block">
<div class="line">to use. For reference check the documentation at</div>
<div class="line"><a class="reference external" href="http://scikit-learn.org/stable/modules/clustering.html">http://scikit-learn.org/stable/modules/clustering.html</a></div>
</div>
<div class="line">params (dict): parameters to be passed to the class when</div>
<div class="line-block">
<div class="line">initialising it. Change depending on the desired</div>
<div class="line">method. Check the documentation for the specific</div>
<div class="line">class.</div>
</div>
</div>
</div>
<div class="line-block">
<div class="line">Returns:</div>
<div class="line-block">
<div class="line">clusters (tuple(list[int],</div>
<div class="line-block">
<div class="line">list[slices])): list of cluster index for each</div>
<div class="line-block">
<div class="line">structure (counting from 1) and</div>
<div class="line">list of slices defining the</div>
<div class="line">clusters as formed by the</div>
<div class="line">requested algorithm.</div>
</div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">load</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.load"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load" title="Link to this definition">#</a></dt>
<dd><p>Load a pickled copy from a given file path</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.save"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save" title="Link to this definition">#</a></dt>
<dd><p>Simply save a pickled copy to a given file path</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection">
<span class="sig-name descname"><span class="pre">save_collection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.save_collection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection" title="Link to this definition">#</a></dt>
<dd><p>Save as pickle the collection bound to this PhylogenCluster.
The calculated genes are also stored in it as arrays for future use.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes">
<span class="sig-name descname"><span class="pre">set_genes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_arrays</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/soprano/analyse/phylogen/phylogenclust.html#PhylogenCluster.set_genes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes" title="Link to this definition">#</a></dt>
<dd><p>Calculate, store and set a list of genes as used for clustering.</p>
<div class="line-block">
<div class="line">Args:</div>
<div class="line-block">
<div class="line">genes (list[soprano.analyse.phylogen.Gene],</div>
<div class="line-block">
<div class="line">file, str): a list of Genes to calculate and store. A path</div>
<div class="line-block">
<div class="line">or open file can also be passed for a .gene</div>
<div class="line">file, from which the values will be loaded.</div>
</div>
</div>
<div class="line">load_arrays (bool): try loading the genes as arrays from the</div>
<div class="line-block">
<div class="line">collection before generating them. Warning:</div>
<div class="line">if there are arrays named like genes but with</div>
<div class="line">different contents this can lead to</div>
<div class="line">unpredictable results.</div>
</div>
</div>
</div>
</dd></dl>

</dd></dl>

</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./_autosummary"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="soprano.analyse.phylogen.metrics.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">soprano.analyse.phylogen.metrics</p>
      </div>
    </a>
    <a class="right-next"
       href="soprano.calculate.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">soprano.calculate</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster._recalc"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster._recalc()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.create_mapping"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.create_mapping()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_cluster_stats"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_cluster_stats()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_distmat"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_distmat()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_elbow_plot"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_elbow_plot()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_matrices()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_matrices_norm"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_matrices_norm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_vectors()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_genome_vectors_norm"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_genome_vectors_norm()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_hier_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_hier_tree"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_hier_tree()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_kmeans_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_kmeans_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_linkage"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_linkage()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_max_cluster_dist"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_max_cluster_dist()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.get_sklearn_clusters"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.get_sklearn_clusters()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.load"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.load()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.save()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.save_collection"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.save_collection()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#soprano.analyse.phylogen.phylogenclust.PhylogenCluster.set_genes"><code class="docutils literal notranslate"><span class="pre">PhylogenCluster.set_genes()</span></code></a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By CCP-NC
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>