
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>soprano.scripts.phylogen &#8212; Soprano</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script >const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
    <script >mermaid.initialize({startOnLoad:true});</script>
    <link rel="shortcut icon" href="../../../_static/favicon.png"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/soprano_logo.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Soprano</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../intro.html">
                    About
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../installation.html">
   Installation
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../tutorials.html">
   Tutorials
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/01-basic_concepts.html">
     Tutorial 1 - Basic concepts: using AtomsCollection objects
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/02-generators_properties_calculators.html">
     Tutorial 2 - Generators, Properties and Calculators
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/03-atomselection_transforms.html">
     Tutorial 3 - AtomSelection and transforms
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/04-clustering.html">
     Tutorial 4 - Clustering analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/05-nmr.html">
     Tutorial 5 - NMR Properties: using Soprano for working with .magres files
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../tutorials/06-defect_calculations.html">
     Tutorial 6 - Soprano for defect calculations
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../submitter.html">
   Job Submitter
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../../api.html">
   API reference
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2 has-children">
    <a class="reference internal" href="../../../_autosummary/soprano.html">
     soprano
    </a>
    <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
    <label for="toctree-checkbox-3">
     <i class="fas fa-chevron-down">
     </i>
    </label>
    <ul>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.analyse.html">
       soprano.analyse
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
      <label for="toctree-checkbox-4">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.analyse.phylogen.html">
         soprano.analyse.phylogen
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.calculate.html">
       soprano.calculate
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
      <label for="toctree-checkbox-5">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.calculate.gulp.html">
         soprano.calculate.gulp
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.calculate.nmr.html">
         soprano.calculate.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.calculate.powder.html">
         soprano.calculate.powder
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.calculate.xrd.html">
         soprano.calculate.xrd
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.collection.html">
       soprano.collection
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
      <label for="toctree-checkbox-6">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.collection.collection.html">
         soprano.collection.collection
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.collection.generate.html">
         soprano.collection.generate
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.data.html">
       soprano.data
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
      <label for="toctree-checkbox-7">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.data.nmr.html">
         soprano.data.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.data.vdw.html">
         soprano.data.vdw
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.hpc.html">
       soprano.hpc
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
      <label for="toctree-checkbox-8">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.hpc.submitter.html">
         soprano.hpc.submitter
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../_autosummary/soprano.measure.html">
       soprano.measure
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.nmr.html">
       soprano.nmr
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
      <label for="toctree-checkbox-9">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.nmr.tensor.html">
         soprano.nmr.tensor
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.nmr.utils.html">
         soprano.nmr.utils
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../_autosummary/soprano.optional.html">
       soprano.optional
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.properties.html">
       soprano.properties
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
      <label for="toctree-checkbox-10">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.atomsproperty.html">
         soprano.properties.atomsproperty
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.basic.html">
         soprano.properties.basic
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.castep.html">
         soprano.properties.castep
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.labeling.html">
         soprano.properties.labeling
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.linkage.html">
         soprano.properties.linkage
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.map.html">
         soprano.properties.map
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.nmr.html">
         soprano.properties.nmr
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.order.html">
         soprano.properties.order
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.symmetry.html">
         soprano.properties.symmetry
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.properties.transform.html">
         soprano.properties.transform
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../_autosummary/soprano.rnd.html">
       soprano.rnd
      </a>
     </li>
     <li class="toctree-l3 has-children">
      <a class="reference internal" href="../../../_autosummary/soprano.scripts.html">
       soprano.scripts
      </a>
      <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
      <label for="toctree-checkbox-11">
       <i class="fas fa-chevron-down">
       </i>
      </label>
      <ul>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.scripts.msaverage.html">
         soprano.scripts.msaverage
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.scripts.phylogen.html">
         soprano.scripts.phylogen
        </a>
       </li>
       <li class="toctree-l4">
        <a class="reference internal" href="../../../_autosummary/soprano.scripts.vasp2cell.html">
         soprano.scripts.vasp2cell
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../_autosummary/soprano.selection.html">
       soprano.selection
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../_autosummary/soprano.utils.html">
       soprano.utils
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../citing.html">
   Citing Soprano
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            A Python library to crack crystals
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/CCP-NC/soprano"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/CCP-NC/soprano/issues/new?title=Issue%20on%20page%20%2F_modules/soprano/scripts/phylogen.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1></h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <h1>Source code for soprano.scripts.phylogen</h1><div class="highlight"><pre>
<span></span><span class="c1"># Soprano - a library to crack crystals! by Simone Sturniolo</span>
<span class="c1"># Copyright (C) 2016 - Science and Technology Facility Council</span>

<span class="c1"># Soprano is free software: you can redistribute it and/or modify</span>
<span class="c1"># it under the terms of the GNU Lesser General Public License as published by</span>
<span class="c1"># the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1"># (at your option) any later version.</span>

<span class="c1"># Soprano is distributed in the hope that it will be useful,</span>
<span class="c1"># but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1"># MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1"># GNU Lesser General Public License for more details.</span>

<span class="c1"># You should have received a copy of the GNU Lesser General Public License</span>
<span class="c1"># along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="sd">&quot;&quot;&quot; A script that allows to run a phylogenetic analysis as a command-line</span>
<span class="sd">tool&quot;&quot;&quot;</span>

<span class="c1"># Python 2-to-3 compatibility code</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">absolute_import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">unicode_literals</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">argparse</span> <span class="k">as</span> <span class="nn">ap</span>
<span class="kn">from</span> <span class="nn">soprano.collection</span> <span class="kn">import</span> <span class="n">AtomsCollection</span>
<span class="kn">from</span> <span class="nn">soprano.properties.basic</span> <span class="kn">import</span> <span class="n">CalcEnergy</span>
<span class="kn">from</span> <span class="nn">soprano.analyse.phylogen</span> <span class="kn">import</span> <span class="n">PhylogenCluster</span><span class="p">,</span> <span class="n">load_genefile</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

    <span class="n">_has_matplotlib</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">_has_matplotlib</span> <span class="o">=</span> <span class="kc">False</span>


<div class="viewcode-block" id="genomes2png"><a class="viewcode-back" href="../../../_autosummary/soprano.scripts.phylogen.html#soprano.scripts.phylogen.genomes2png">[docs]</a><span class="k">def</span> <span class="nf">genomes2png</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">outf</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">90</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Produce a PNG plot of a series of genomes vectors&quot;&quot;&quot;</span>

    <span class="n">genomes</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">pC</span><span class="o">.</span><span class="n">get_genome_vectors_norm</span><span class="p">()</span>
    <span class="n">h</span><span class="p">,</span> <span class="n">w</span> <span class="o">=</span> <span class="n">genomes</span><span class="o">.</span><span class="n">shape</span>
    <span class="c1"># Now build the labels</span>
    <span class="n">xlabels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">legend</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
    <span class="n">ylabels</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="c1"># Calculate the size</span>
    <span class="n">figw</span> <span class="o">=</span> <span class="n">w</span> <span class="o">*</span> <span class="mf">90.0</span> <span class="o">+</span> <span class="mf">200.0</span>
    <span class="n">figh</span> <span class="o">=</span> <span class="n">h</span> <span class="o">*</span> <span class="mf">50.0</span> <span class="o">+</span> <span class="mf">100.0</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">figw</span> <span class="o">/</span> <span class="n">dpi</span><span class="p">,</span> <span class="n">figh</span> <span class="o">/</span> <span class="n">dpi</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span>
        <span class="n">left</span><span class="o">=</span><span class="mf">150.0</span> <span class="o">/</span> <span class="n">figw</span><span class="p">,</span>
        <span class="n">right</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="mf">50.0</span> <span class="o">/</span> <span class="n">figw</span><span class="p">,</span>
        <span class="n">bottom</span><span class="o">=</span><span class="mf">50.0</span> <span class="o">/</span> <span class="n">figh</span><span class="p">,</span>
        <span class="n">top</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="mf">50.0</span> <span class="o">/</span> <span class="n">figh</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">pcolor</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;jet&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">labelsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">w</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">w</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">xlabels</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">h</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">h</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">ylabels</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">outf</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span></div>


<div class="viewcode-block" id="tree2json"><a class="viewcode-back" href="../../../_autosummary/soprano.scripts.phylogen.html#soprano.scripts.phylogen.tree2json">[docs]</a><span class="k">def</span> <span class="nf">tree2json</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">h</span><span class="o">=</span><span class="mi">500</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Produce a dictionary representation of a Scipy cluster tree object</span>
<span class="sd">    suitable to be saved as JSON&quot;&quot;&quot;</span>

    <span class="c1"># First find absolute minimum and maximum energy (for hue)</span>

    <span class="n">struct_Es</span> <span class="o">=</span> <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="p">)</span>

    <span class="n">Emin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">struct_Es</span><span class="p">)</span>
    <span class="n">Emax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">struct_Es</span><span class="p">)</span>

    <span class="c1"># Create the master node by using the tree root, then visit recursively</span>

    <span class="k">def</span> <span class="nf">visit_tree</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>

        <span class="n">datastr</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">get_count</span><span class="p">(),</span> <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">(),</span> <span class="s2">&quot;children&quot;</span><span class="p">:</span> <span class="p">[]}</span>

        <span class="c1"># Add label text if possible</span>

        <span class="k">for</span> <span class="n">br</span> <span class="ow">in</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">left</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">right</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">br</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">visit_tree</span><span class="p">(</span><span class="n">br</span><span class="p">))</span>

        <span class="c1"># Now the hue and the struct name. If this is a leaf, just use the</span>
        <span class="c1"># structure value</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">Emax</span> <span class="o">&gt;</span> <span class="n">Emin</span><span class="p">:</span>
                <span class="n">hue</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">Emax</span> <span class="o">-</span> <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">()]))</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">Emax</span> <span class="o">-</span> <span class="n">Emin</span><span class="p">)</span>
                    <span class="o">*</span> <span class="mf">240.0</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hue</span> <span class="o">=</span> <span class="mf">120.0</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">()]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Otherwise, an average</span>
            <span class="n">hue</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">]:</span>
                <span class="n">hue</span> <span class="o">+=</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;hue&quot;</span><span class="p">]</span>
                <span class="n">struct_name</span> <span class="o">+=</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;struct_name&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">hue</span> <span class="o">/=</span> <span class="nb">len</span><span class="p">(</span><span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">])</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="n">struct_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>  <span class="c1"># Remove last newline and comma</span>

        <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;struct_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">struct_name</span>
        <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;hue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hue</span>

        <span class="k">return</span> <span class="n">datastr</span>

    <span class="k">return</span> <span class="n">visit_tree</span><span class="p">(</span><span class="n">pC</span><span class="o">.</span><span class="n">get_hier_tree</span><span class="p">())</span></div>


<span class="c1"># Entry point</span>
<span class="k">def</span> <span class="nf">__main__</span><span class="p">():</span>
    <span class="c1"># So, first parse the command line arguments</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ap</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="c1"># Main argument</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;seedname&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Seedname of the job&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;input_files&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Files to load and analyse&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Optional arguments</span>
    <span class="c1"># Boolean args</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-pkl&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Input is one or more AtomsCollections&quot;</span> <span class="s2">&quot; in PICKLE format&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-tree&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Create JSON tree file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-distmat&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save a distance matrix file&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-forcemat&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Create JSON distance matrix file for force &quot;</span> <span class="s2">&quot;layout&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-genomes&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save genomes text file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-gene_png&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save genomes png file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-gene_clamp&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Clamp genes between 0 and 1&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Value args</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-n&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of lowest energy structures to keep&quot;</span>
    <span class="p">)</span>
    <span class="c1"># parser.add_argument(&#39;-Z&#39;, type=int, default=None,</span>
    <span class="c1"># help=&quot;Expected numbers of molecules in cell (discard structures &quot;</span>
    <span class="c1"># &quot;that don&#39;t agree)&quot;)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-corrmat_minforce&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Correlation matrix minimum force&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-corrmat_maxforce&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Correlation matrix maximum force&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-savepkl&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Directory to save collection as pickle file&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># First, parse the gene list</span>
    <span class="n">gene_list</span> <span class="o">=</span> <span class="n">load_genefile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;.gene&quot;</span><span class="p">)</span>

    <span class="c1"># Then create an AtomsCollection</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">pkl</span><span class="p">:</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">AtomsCollection</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">input_files</span><span class="p">,</span>
            <span class="n">info</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">seedname</span><span class="p">},</span>
            <span class="n">cell_reduce</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">progress</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">AtomsCollection</span><span class="p">([])</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">input_files</span><span class="p">:</span>
            <span class="n">aC</span> <span class="o">+=</span> <span class="n">AtomsCollection</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="c1"># Take only n lowest energies if required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Sort by energy</span>
        <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">store_array</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">sorted_byarray</span><span class="p">(</span><span class="s2">&quot;calc_energy&quot;</span><span class="p">)[:</span> <span class="n">args</span><span class="o">.</span><span class="n">n</span><span class="p">]</span>

    <span class="c1"># discard_Z = args.Z is not None</span>

    <span class="c1"># Calculate genomes</span>
    <span class="n">nrange</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_clamp</span> <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">ndist</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_clamp</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">pC</span> <span class="o">=</span> <span class="n">PhylogenCluster</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="p">[],</span> <span class="n">norm_range</span><span class="o">=</span><span class="n">nrange</span><span class="p">,</span> <span class="n">norm_dist</span><span class="o">=</span><span class="n">ndist</span><span class="p">)</span>
    <span class="n">pC</span><span class="o">.</span><span class="n">set_genes</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">load_arrays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">savepkl</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Save the structure</span>
        <span class="n">pC</span><span class="o">.</span><span class="n">save_collection</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">savepkl</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;.pkl&quot;</span><span class="p">))</span>

    <span class="c1"># And now for the output</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genomes</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_genomes.dat&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">genvec</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">pC</span><span class="o">.</span><span class="n">get_genome_vectors</span><span class="p">()</span>
            <span class="c1"># Create a header</span>
            <span class="n">header</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Structure</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
                        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">legend</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genvec</span><span class="p">):</span>
                    <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                        <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
                        <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gen</span><span class="p">])</span>
                        <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="c1"># Someone has no name!</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;Not all structures have a name, &quot;</span>
                    <span class="s2">&quot;can&#39;t print out genomes as table&quot;</span>
                <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_png</span> <span class="ow">and</span> <span class="n">_has_matplotlib</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_genomes.png&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">genomes2png</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">ofile</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_tree.json&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">tree2json</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">),</span> <span class="n">ofile</span><span class="p">)</span>
</pre></div>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By CCP-NC<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>