
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>soprano.scripts.phylogen &#8212; Soprano</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=87e54e7c" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=5eb1bb1e"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
    <script>mermaid.initialize({startOnLoad:true});</script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/soprano/scripts/phylogen';</script>
    <link rel="icon" href="../../../_static/favicon.png"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../intro.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/soprano_logo.svg" class="logo__image only-light" alt="Soprano - Home"/>
    <script>document.write(`<img src="../../../_static/soprano_logo.svg" class="logo__image only-dark" alt="Soprano - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../intro.html">
                    About
                </a>
            </li>
        </ul>
        <ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../tutorials.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/01-basic_concepts.html">Tutorial 1 - Basic concepts: using AtomsCollection objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/02-generators_properties_calculators.html">Tutorial 2 - Generators, Properties and Calculators</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/03-atomselection_transforms.html">Tutorial 3 - AtomSelection and transforms</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/04-clustering.html">Tutorial 4 - Clustering analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/05-nmr.html">Tutorial 5 - NMR Properties: using Soprano for working with .magres files</a></li>




<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/06-defect_calculations.html">Tutorial 6 - Soprano for defect calculations</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../../submitter.html">Job Submitter</a></li>

<li class="toctree-l1 has-children"><a class="reference internal" href="../../../api.html">API reference</a><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../_autosummary/soprano.html">soprano</a><ul>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.analyse.html">soprano.analyse</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.analyse.phylogen.html">soprano.analyse.phylogen</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.calculate.html">soprano.calculate</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.calculate.gulp.html">soprano.calculate.gulp</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.calculate.nmr.html">soprano.calculate.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.calculate.powder.html">soprano.calculate.powder</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.calculate.xrd.html">soprano.calculate.xrd</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.collection.html">soprano.collection</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.collection.collection.html">soprano.collection.collection</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.collection.generate.html">soprano.collection.generate</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.data.html">soprano.data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.data.nmr.html">soprano.data.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.data.vdw.html">soprano.data.vdw</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.hpc.html">soprano.hpc</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.hpc.submitter.html">soprano.hpc.submitter</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="../../../_autosummary/soprano.measure.html">soprano.measure</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.nmr.html">soprano.nmr</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.nmr.tensor.html">soprano.nmr.tensor</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.nmr.utils.html">soprano.nmr.utils</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="../../../_autosummary/soprano.optional.html">soprano.optional</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.properties.html">soprano.properties</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.atomsproperty.html">soprano.properties.atomsproperty</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.basic.html">soprano.properties.basic</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.castep.html">soprano.properties.castep</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.labeling.html">soprano.properties.labeling</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.linkage.html">soprano.properties.linkage</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.map.html">soprano.properties.map</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.nmr.html">soprano.properties.nmr</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.order.html">soprano.properties.order</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.symmetry.html">soprano.properties.symmetry</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.properties.transform.html">soprano.properties.transform</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="../../../_autosummary/soprano.rnd.html">soprano.rnd</a></li>
<li class="toctree-l3 has-children"><a class="reference internal" href="../../../_autosummary/soprano.scripts.html">soprano.scripts</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.scripts.msaverage.html">soprano.scripts.msaverage</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.scripts.phylogen.html">soprano.scripts.phylogen</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../_autosummary/soprano.scripts.vasp2cell.html">soprano.scripts.vasp2cell</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
<li class="toctree-l3"><a class="reference internal" href="../../../_autosummary/soprano.selection.html">soprano.selection</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../_autosummary/soprano.utils.html">soprano.utils</a></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>
</ul><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary>
</details></li>

<li class="toctree-l1"><a class="reference internal" href="../../../citing.html">Citing Soprano</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/CCP-NC/soprano" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/CCP-NC/soprano/issues/new?title=Issue%20on%20page%20%2F_modules/soprano/scripts/phylogen.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for soprano.scripts.phylogen</h1><div class="highlight"><pre>
<span></span><span class="c1"># Soprano - a library to crack crystals! by Simone Sturniolo</span>
<span class="c1"># Copyright (C) 2016 - Science and Technology Facility Council</span>

<span class="c1"># Soprano is free software: you can redistribute it and/or modify</span>
<span class="c1"># it under the terms of the GNU Lesser General Public License as published by</span>
<span class="c1"># the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1"># (at your option) any later version.</span>

<span class="c1"># Soprano is distributed in the hope that it will be useful,</span>
<span class="c1"># but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1"># MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1"># GNU Lesser General Public License for more details.</span>

<span class="c1"># You should have received a copy of the GNU Lesser General Public License</span>
<span class="c1"># along with this program.  If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>


<span class="sd">&quot;&quot;&quot; A script that allows to run a phylogenetic analysis as a command-line</span>
<span class="sd">tool&quot;&quot;&quot;</span>

<span class="c1"># Python 2-to-3 compatibility code</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">absolute_import</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">division</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">unicode_literals</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">argparse</span> <span class="k">as</span> <span class="nn">ap</span>
<span class="kn">from</span> <span class="nn">soprano.collection</span> <span class="kn">import</span> <span class="n">AtomsCollection</span>
<span class="kn">from</span> <span class="nn">soprano.properties.basic</span> <span class="kn">import</span> <span class="n">CalcEnergy</span>
<span class="kn">from</span> <span class="nn">soprano.analyse.phylogen</span> <span class="kn">import</span> <span class="n">PhylogenCluster</span><span class="p">,</span> <span class="n">load_genefile</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

    <span class="n">_has_matplotlib</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">_has_matplotlib</span> <span class="o">=</span> <span class="kc">False</span>


<div class="viewcode-block" id="genomes2png">
<a class="viewcode-back" href="../../../_autosummary/soprano.scripts.phylogen.html#soprano.scripts.phylogen.genomes2png">[docs]</a>
<span class="k">def</span> <span class="nf">genomes2png</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">outf</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">90</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Produce a PNG plot of a series of genomes vectors&quot;&quot;&quot;</span>

    <span class="n">genomes</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">pC</span><span class="o">.</span><span class="n">get_genome_vectors_norm</span><span class="p">()</span>
    <span class="n">h</span><span class="p">,</span> <span class="n">w</span> <span class="o">=</span> <span class="n">genomes</span><span class="o">.</span><span class="n">shape</span>
    <span class="c1"># Now build the labels</span>
    <span class="n">xlabels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">legend</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span>
    <span class="n">ylabels</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">all</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">s</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
    <span class="c1"># Calculate the size</span>
    <span class="n">figw</span> <span class="o">=</span> <span class="n">w</span> <span class="o">*</span> <span class="mf">90.0</span> <span class="o">+</span> <span class="mf">200.0</span>
    <span class="n">figh</span> <span class="o">=</span> <span class="n">h</span> <span class="o">*</span> <span class="mf">50.0</span> <span class="o">+</span> <span class="mf">100.0</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="n">figw</span> <span class="o">/</span> <span class="n">dpi</span><span class="p">,</span> <span class="n">figh</span> <span class="o">/</span> <span class="n">dpi</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span>
        <span class="n">left</span><span class="o">=</span><span class="mf">150.0</span> <span class="o">/</span> <span class="n">figw</span><span class="p">,</span>
        <span class="n">right</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="mf">50.0</span> <span class="o">/</span> <span class="n">figw</span><span class="p">,</span>
        <span class="n">bottom</span><span class="o">=</span><span class="mf">50.0</span> <span class="o">/</span> <span class="n">figh</span><span class="p">,</span>
        <span class="n">top</span><span class="o">=</span><span class="mf">1.0</span> <span class="o">-</span> <span class="mf">50.0</span> <span class="o">/</span> <span class="n">figh</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">pcolor</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">edgecolors</span><span class="o">=</span><span class="s2">&quot;black&quot;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;jet&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">labelsize</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">w</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">w</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">xlabels</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">h</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">h</span><span class="p">)</span> <span class="o">+</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">ylabels</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">outf</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;png&quot;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="n">dpi</span><span class="p">)</span></div>



<div class="viewcode-block" id="tree2json">
<a class="viewcode-back" href="../../../_autosummary/soprano.scripts.phylogen.html#soprano.scripts.phylogen.tree2json">[docs]</a>
<span class="k">def</span> <span class="nf">tree2json</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">w</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">h</span><span class="o">=</span><span class="mi">500</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Produce a dictionary representation of a Scipy cluster tree object</span>
<span class="sd">    suitable to be saved as JSON&quot;&quot;&quot;</span>

    <span class="c1"># First find absolute minimum and maximum energy (for hue)</span>

    <span class="n">struct_Es</span> <span class="o">=</span> <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="p">)</span>

    <span class="n">Emin</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">struct_Es</span><span class="p">)</span>
    <span class="n">Emax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">struct_Es</span><span class="p">)</span>

    <span class="c1"># Create the master node by using the tree root, then visit recursively</span>

    <span class="k">def</span> <span class="nf">visit_tree</span><span class="p">(</span><span class="n">node</span><span class="p">):</span>

        <span class="n">datastr</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;value&quot;</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">get_count</span><span class="p">(),</span> <span class="s2">&quot;id&quot;</span><span class="p">:</span> <span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">(),</span> <span class="s2">&quot;children&quot;</span><span class="p">:</span> <span class="p">[]}</span>

        <span class="c1"># Add label text if possible</span>

        <span class="k">for</span> <span class="n">br</span> <span class="ow">in</span> <span class="p">(</span><span class="n">node</span><span class="o">.</span><span class="n">left</span><span class="p">,</span> <span class="n">node</span><span class="o">.</span><span class="n">right</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">br</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">visit_tree</span><span class="p">(</span><span class="n">br</span><span class="p">))</span>

        <span class="c1"># Now the hue and the struct name. If this is a leaf, just use the</span>
        <span class="c1"># structure value</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">Emax</span> <span class="o">&gt;</span> <span class="n">Emin</span><span class="p">:</span>
                <span class="n">hue</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="p">(</span><span class="n">Emax</span> <span class="o">-</span> <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">()]))</span>
                    <span class="o">/</span> <span class="p">(</span><span class="n">Emax</span> <span class="o">-</span> <span class="n">Emin</span><span class="p">)</span>
                    <span class="o">*</span> <span class="mf">240.0</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">hue</span> <span class="o">=</span> <span class="mf">120.0</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">node</span><span class="o">.</span><span class="n">get_id</span><span class="p">()]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Otherwise, an average</span>
            <span class="n">hue</span> <span class="o">=</span> <span class="mf">0.0</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">]:</span>
                <span class="n">hue</span> <span class="o">+=</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;hue&quot;</span><span class="p">]</span>
                <span class="n">struct_name</span> <span class="o">+=</span> <span class="n">c</span><span class="p">[</span><span class="s2">&quot;struct_name&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;,</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">hue</span> <span class="o">/=</span> <span class="nb">len</span><span class="p">(</span><span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;children&quot;</span><span class="p">])</span>
            <span class="n">struct_name</span> <span class="o">=</span> <span class="n">struct_name</span><span class="p">[:</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>  <span class="c1"># Remove last newline and comma</span>

        <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;struct_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">struct_name</span>
        <span class="n">datastr</span><span class="p">[</span><span class="s2">&quot;hue&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hue</span>

        <span class="k">return</span> <span class="n">datastr</span>

    <span class="k">return</span> <span class="n">visit_tree</span><span class="p">(</span><span class="n">pC</span><span class="o">.</span><span class="n">get_hier_tree</span><span class="p">())</span></div>



<span class="c1"># Entry point</span>
<span class="k">def</span> <span class="nf">__main__</span><span class="p">():</span>
    <span class="c1"># So, first parse the command line arguments</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">ap</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
    <span class="c1"># Main argument</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;seedname&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Seedname of the job&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;input_files&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">nargs</span><span class="o">=</span><span class="s2">&quot;+&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Files to load and analyse&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Optional arguments</span>
    <span class="c1"># Boolean args</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-pkl&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Input is one or more AtomsCollections&quot;</span> <span class="s2">&quot; in PICKLE format&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-tree&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Create JSON tree file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-distmat&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save a distance matrix file&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-forcemat&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Create JSON distance matrix file for force &quot;</span> <span class="s2">&quot;layout&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-genomes&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save genomes text file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-gene_png&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Save genomes png file&quot;</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-gene_clamp&quot;</span><span class="p">,</span>
        <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Clamp genes between 0 and 1&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="c1"># Value args</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-n&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Number of lowest energy structures to keep&quot;</span>
    <span class="p">)</span>
    <span class="c1"># parser.add_argument(&#39;-Z&#39;, type=int, default=None,</span>
    <span class="c1"># help=&quot;Expected numbers of molecules in cell (discard structures &quot;</span>
    <span class="c1"># &quot;that don&#39;t agree)&quot;)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-corrmat_minforce&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Correlation matrix minimum force&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-corrmat_maxforce&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="mf">1.0</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Correlation matrix maximum force&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
        <span class="s2">&quot;-savepkl&quot;</span><span class="p">,</span>
        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
        <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Directory to save collection as pickle file&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

    <span class="c1"># First, parse the gene list</span>
    <span class="n">gene_list</span> <span class="o">=</span> <span class="n">load_genefile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;.gene&quot;</span><span class="p">)</span>

    <span class="c1"># Then create an AtomsCollection</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">pkl</span><span class="p">:</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">AtomsCollection</span><span class="p">(</span>
            <span class="n">args</span><span class="o">.</span><span class="n">input_files</span><span class="p">,</span>
            <span class="n">info</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="n">args</span><span class="o">.</span><span class="n">seedname</span><span class="p">},</span>
            <span class="n">cell_reduce</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">progress</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">AtomsCollection</span><span class="p">([])</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">input_files</span><span class="p">:</span>
            <span class="n">aC</span> <span class="o">+=</span> <span class="n">AtomsCollection</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>

    <span class="c1"># Take only n lowest energies if required</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">n</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Sort by energy</span>
        <span class="n">CalcEnergy</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">store_array</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">aC</span> <span class="o">=</span> <span class="n">aC</span><span class="o">.</span><span class="n">sorted_byarray</span><span class="p">(</span><span class="s2">&quot;calc_energy&quot;</span><span class="p">)[:</span> <span class="n">args</span><span class="o">.</span><span class="n">n</span><span class="p">]</span>

    <span class="c1"># discard_Z = args.Z is not None</span>

    <span class="c1"># Calculate genomes</span>
    <span class="n">nrange</span> <span class="o">=</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_clamp</span> <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">ndist</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_clamp</span> <span class="k">else</span> <span class="kc">None</span>
    <span class="n">pC</span> <span class="o">=</span> <span class="n">PhylogenCluster</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="p">[],</span> <span class="n">norm_range</span><span class="o">=</span><span class="n">nrange</span><span class="p">,</span> <span class="n">norm_dist</span><span class="o">=</span><span class="n">ndist</span><span class="p">)</span>
    <span class="n">pC</span><span class="o">.</span><span class="n">set_genes</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">load_arrays</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">savepkl</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Save the structure</span>
        <span class="n">pC</span><span class="o">.</span><span class="n">save_collection</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">savepkl</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;.pkl&quot;</span><span class="p">))</span>

    <span class="c1"># And now for the output</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genomes</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_genomes.dat&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">genvec</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="n">pC</span><span class="o">.</span><span class="n">get_genome_vectors</span><span class="p">()</span>
            <span class="c1"># Create a header</span>
            <span class="n">header</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Structure</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="p">[</span>
                        <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">l</span><span class="p">[</span><span class="mi">1</span><span class="p">])])</span>
                        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">legend</span>
                    <span class="p">]</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">header</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">gen</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">genvec</span><span class="p">):</span>
                    <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="n">aC</span><span class="o">.</span><span class="n">structures</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
                        <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span>
                        <span class="o">+</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gen</span><span class="p">])</span>
                        <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="c1"># Someone has no name!</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                    <span class="s2">&quot;Not all structures have a name, &quot;</span>
                    <span class="s2">&quot;can&#39;t print out genomes as table&quot;</span>
                <span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">gene_png</span> <span class="ow">and</span> <span class="n">_has_matplotlib</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_genomes.png&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">genomes2png</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">,</span> <span class="n">ofile</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">tree</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">seedname</span> <span class="o">+</span> <span class="s2">&quot;_tree.json&quot;</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">tree2json</span><span class="p">(</span><span class="n">aC</span><span class="p">,</span> <span class="n">pC</span><span class="p">),</span> <span class="n">ofile</span><span class="p">)</span>
</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By CCP-NC
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>